{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.1 Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting biopython\n",
      "  Downloading biopython-1.83-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
      "Requirement already satisfied: numpy in /usr/local/lib/python3.9/site-packages (from biopython) (1.26.4)\n",
      "Downloading biopython-1.83-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m19.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: biopython\n",
      "Successfully installed biopython-1.83\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/osheakes/Research_Project_MMM/Fasta\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1.0 Viral Precursor Blast "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 Viral Blast test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test run of one Accession to ensure command runs as expected\n",
    "!blastp -remote -query input_proteins.fa -db genomic/Viruses/Viral_Protein_Sequences -evalue 0.06 -outfmt 7 -out output.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Blast in fmt 5 format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping O96011.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O96011_blast_results.txt\n",
      "Skipping Q07812.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q07812_blast_results.txt\n",
      "Skipping Q8NFJ6.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q8NFJ6_blast_results.txt\n",
      "Skipping P07492.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P07492_blast_results.txt\n",
      "Skipping Q15303.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q15303_blast_results.txt\n",
      "Skipping Q15726.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q15726_blast_results.txt\n",
      "Skipping O00910.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O00910_blast_results.txt\n",
      "Skipping P38398.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P38398_blast_results.txt\n",
      "Skipping P06881.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P06881_blast_results.txt\n",
      "Skipping P46663.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P46663_blast_results.txt\n",
      "Skipping P32247.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P32247_blast_results.txt\n",
      "Skipping P60484.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P60484_blast_results.txt\n",
      "Skipping P20366.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P20366_blast_results.txt\n",
      "Skipping P35318.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P35318_blast_results.txt\n",
      "Skipping P12872.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P12872_blast_results.txt\n",
      "Skipping P10645.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P10645_blast_results.txt\n",
      "Skipping O95760.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O95760_blast_results.txt\n",
      "Skipping P01344.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01344_blast_results.txt\n",
      "Skipping O14879.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O14879_blast_results.txt\n",
      "Skipping O43745.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O43745_blast_results.txt\n",
      "Skipping Q92968.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q92968_blast_results.txt\n",
      "Skipping O95390.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O95390_blast_results.txt\n",
      "Skipping P02533.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P02533_blast_results.txt\n",
      "Skipping Q01629.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q01629_blast_results.txt\n",
      "Skipping P0C0P6.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P0C0P6_blast_results.txt\n",
      "Skipping Q00534.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q00534_blast_results.txt\n",
      "Skipping P20592.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P20592_blast_results.txt\n",
      "Skipping P06493.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P06493_blast_results.txt\n",
      "Skipping P06400.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P06400_blast_results.txt\n",
      "Skipping Q03252.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q03252_blast_results.txt\n",
      "Skipping P24941.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P24941_blast_results.txt\n",
      "Skipping Q13519.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q13519_blast_results.txt\n",
      "Skipping P01574.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01574_blast_results.txt\n",
      "Skipping P19883.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P19883_blast_results.txt\n",
      "Skipping P01282.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01282_blast_results.txt\n",
      "Skipping Q9Y5Q6.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9Y5Q6_blast_results.txt\n",
      "Skipping Q8IUX4.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q8IUX4_blast_results.txt\n",
      "Skipping P55265.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P55265_blast_results.txt\n",
      "Skipping Q99988.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q99988_blast_results.txt\n",
      "Skipping P48645.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P48645_blast_results.txt\n",
      "Skipping P06396.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P06396_blast_results.txt\n",
      "Skipping P01966.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01966_blast_results.txt\n",
      "Skipping Q96BS2.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q96BS2_blast_results.txt\n",
      "Skipping P47898.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P47898_blast_results.txt\n",
      "Skipping P42226.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P42226_blast_results.txt\n",
      "Skipping P09914.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P09914_blast_results.txt\n",
      "Skipping O75192.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O75192_blast_results.txt\n",
      "Skipping O95399.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O95399_blast_results.txt\n",
      "Skipping P11802.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P11802_blast_results.txt\n",
      "Skipping P08949.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P08949_blast_results.txt\n",
      "Skipping Q01453.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q01453_blast_results.txt\n",
      "Skipping Q10589.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q10589_blast_results.txt\n",
      "Skipping Q969E3.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q969E3_blast_results.txt\n",
      "Skipping Q14790.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q14790_blast_results.txt\n",
      "Skipping A0A097EV82.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/A0A097EV82_blast_results.txt\n",
      "Skipping P55089.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P55089_blast_results.txt\n",
      "Skipping O75381.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O75381_blast_results.txt\n",
      "Skipping Q9P0W0.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9P0W0_blast_results.txt\n",
      "Skipping Q9H293.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9H293_blast_results.txt\n",
      "Skipping Q00535.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q00535_blast_results.txt\n",
      "Skipping P43490.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P43490_blast_results.txt\n",
      "Skipping Q9BTW9.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9BTW9_blast_results.txt\n",
      "Skipping P42224.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P42224_blast_results.txt\n",
      "Skipping Q00526.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q00526_blast_results.txt\n",
      "Skipping Q9HC23.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9HC23_blast_results.txt\n",
      "Skipping P01241.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01241_blast_results.txt\n",
      "Skipping Q16611.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q16611_blast_results.txt\n",
      "Skipping P56199.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P56199_blast_results.txt\n",
      "Skipping P01258.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01258_blast_results.txt\n",
      "Skipping Q8NG41.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q8NG41_blast_results.txt\n",
      "Skipping P10082.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P10082_blast_results.txt\n",
      "Skipping Q7Z2W4.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q7Z2W4_blast_results.txt\n",
      "Skipping P19525.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P19525_blast_results.txt\n",
      "Skipping O43808.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O43808_blast_results.txt\n",
      "Skipping Q96RP3.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q96RP3_blast_results.txt\n",
      "Skipping P01566.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01566_blast_results.txt\n",
      "Skipping P01568.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01568_blast_results.txt\n",
      "Skipping P80303.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P80303_blast_results.txt\n",
      "Skipping O15130.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O15130_blast_results.txt\n",
      "Skipping P01019.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01019_blast_results.txt\n",
      "Skipping Q99653.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q99653_blast_results.txt\n",
      "Skipping P01570.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01570_blast_results.txt\n",
      "Skipping P30990.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P30990_blast_results.txt\n",
      "Skipping P23582.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P23582_blast_results.txt\n",
      "Skipping P42574.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P42574_blast_results.txt\n",
      "Skipping P16860.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P16860_blast_results.txt\n",
      "Skipping Q14765.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q14765_blast_results.txt\n",
      "Skipping Q9Y3Z3.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9Y3Z3_blast_results.txt\n",
      "Skipping P01579.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01579_blast_results.txt\n",
      "Skipping Q02388.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q02388_blast_results.txt\n",
      "Skipping Q8N729.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q8N729_blast_results.txt\n",
      "Skipping O95786.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/O95786_blast_results.txt\n",
      "Skipping P13164.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P13164_blast_results.txt\n",
      "Skipping P18509.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P18509_blast_results.txt\n",
      "Skipping Q12965.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q12965_blast_results.txt\n",
      "Skipping P01160.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01160_blast_results.txt\n",
      "Skipping P81605.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P81605_blast_results.txt\n",
      "Skipping P09912.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P09912_blast_results.txt\n",
      "Skipping Q13325.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q13325_blast_results.txt\n",
      "Skipping P20591.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P20591_blast_results.txt\n",
      "Skipping P52630.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P52630_blast_results.txt\n",
      "Skipping P01210.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01210_blast_results.txt\n",
      "Skipping P40763.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P40763_blast_results.txt\n",
      "Skipping P01213.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01213_blast_results.txt\n",
      "Skipping Q01628.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q01628_blast_results.txt\n",
      "Skipping P23142.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P23142_blast_results.txt\n",
      "Skipping Q9UHF0.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9UHF0_blast_results.txt\n",
      "Skipping P51587.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P51587_blast_results.txt\n",
      "Skipping P15502.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P15502_blast_results.txt\n",
      "Skipping P19838.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P19838_blast_results.txt\n",
      "Skipping P04637.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P04637_blast_results.txt\n",
      "Skipping Q5H8A3.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q5H8A3_blast_results.txt\n",
      "Skipping P55211.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P55211_blast_results.txt\n",
      "Skipping P25445.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P25445_blast_results.txt\n",
      "Skipping Q9Y5Y5.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/Q9Y5Y5_blast_results.txt\n",
      "Skipping P01303.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01303_blast_results.txt\n",
      "Skipping P55210.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P55210_blast_results.txt\n",
      "Skipping P01225.fasta - Output file already exists: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/P01225_blast_results.txt\n",
      "Total execution time: 1 minutes 0.27 seconds\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "\n",
    "# Record start time\n",
    "start_time = time.time()\n",
    "\n",
    "# fasta inputs\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta\"\n",
    "\n",
    "# Directory for results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/\"\n",
    "\n",
    "# BLAST parameters\n",
    "blast_db = \"genomic/Viruses/Viral_Protein_Sequences\"\n",
    "evalue = \"0.06\"\n",
    "outfmt = \"5\"  \n",
    "blast_remote = \"-remote\"  \n",
    "\n",
    "# 5min interval for logbook\n",
    "print_interval = 300 \n",
    "\n",
    "# Set initial time for printing logbook\n",
    "next_print_time = start_time + print_interval\n",
    "\n",
    "# Select fasta's in directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):  \n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        output_file = os.path.splitext(filename)[0] + \"_blast_results.txt\"\n",
    "        output_path = os.path.join(blast_output_directory, output_file)\n",
    "        \n",
    "        # Check if output file already exists\n",
    "        if os.path.exists(output_path):\n",
    "            print(f\"Skipping {filename} - Output file already exists: {output_path}\")\n",
    "            continue\n",
    "        \n",
    "        # Record start time for each file\n",
    "        file_start_time = time.time()\n",
    "        \n",
    "        # BLAST command \n",
    "        command = f\"blastp {blast_remote} -query {fasta_file} -db {blast_db} -evalue {evalue} -outfmt {outfmt} -out {output_path}\"\n",
    "        \n",
    "        # Execute BLAST command\n",
    "        os.system(command)\n",
    "        \n",
    "        # Calculate time taken for BLAST search on current file\n",
    "        file_duration = time.time() - file_start_time\n",
    "        \n",
    "        #  to print logbook\n",
    "        if time.time() >= next_print_time:\n",
    "            # Print logbook message\n",
    "            print(f\"BLAST for {filename} started at {time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime(file_start_time))} and completed in {file_duration//60:.0f} minutes {file_duration%60:.2f} seconds. Results saved in {output_path}\")\n",
    "            \n",
    "            # Update next print time\n",
    "            next_print_time += print_interval\n",
    "        \n",
    "# Calculate total run time\n",
    "total_duration = time.time() - start_time\n",
    "\n",
    "# Print total execution time\n",
    "print(f\"Total execution time: {total_duration//60:.0f} minutes {total_duration%60:.2f} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of FASTA files in directory '/home/osheakes/Research_Project_MMM/Fasta/': 119\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Path to the directory containing the FASTA files\n",
    "directory_path = \"/home/osheakes/Research_Project_MMM/Fasta/\"\n",
    "\n",
    "# Count the number of FASTA files in the directory\n",
    "fasta_files = [file for file in os.listdir(directory_path) if file.endswith(\".fasta\")]\n",
    "num_fasta_files = len(fasta_files)\n",
    "\n",
    "print(f\"Number of FASTA files in directory '{directory_path}': {num_fasta_files}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Create script to parse data out of blast results files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file: P12872_blast_results.txt\n",
      "Ignoring file: P12872_blast_results.txt - No hits found\n",
      "Processing file: P56199_blast_results.txt\n",
      "Processing file: P81605_blast_results.txt\n",
      "Ignoring file: P81605_blast_results.txt - No hits found\n",
      "Processing file: P01579_blast_results.txt\n",
      "Ignoring file: P01579_blast_results.txt - No hits found\n",
      "Processing file: P18509_blast_results.txt\n",
      "Ignoring file: P18509_blast_results.txt - No hits found\n",
      "Processing file: P01258_blast_results.txt\n",
      "Ignoring file: P01258_blast_results.txt - No hits found\n",
      "Processing file: P01570_blast_results.txt\n",
      "Ignoring file: P01570_blast_results.txt - No hits found\n",
      "Processing file: P23582_blast_results.txt\n",
      "Ignoring file: P23582_blast_results.txt - No hits found\n",
      "Processing file: P24941_blast_results.txt\n",
      "Processing file: P19838_blast_results.txt\n",
      "Processing file: O95390_blast_results.txt\n",
      "Processing file: Q9HC23_blast_results.txt\n",
      "Ignoring file: Q9HC23_blast_results.txt - No hits found\n",
      "Processing file: Q8NFJ6_blast_results.txt\n",
      "Processing file: O95786_blast_results.txt\n",
      "Processing file: P20366_blast_results.txt\n",
      "Ignoring file: P20366_blast_results.txt - No hits found\n",
      "Processing file: P01568_blast_results.txt\n",
      "Ignoring file: P01568_blast_results.txt - No hits found\n",
      "Processing file: O95760_blast_results.txt\n",
      "Ignoring file: O95760_blast_results.txt - No hits found\n",
      "Processing file: P06881_blast_results.txt\n",
      "Ignoring file: P06881_blast_results.txt - No hits found\n",
      "Processing file: Q5H8A3_blast_results.txt\n",
      "Ignoring file: Q5H8A3_blast_results.txt - No hits found\n",
      "Processing file: O43808_blast_results.txt\n",
      "Processing file: O14879_blast_results.txt\n",
      "Processing file: P01966_blast_results.txt\n",
      "Processing file: P01282_blast_results.txt\n",
      "Ignoring file: P01282_blast_results.txt - No hits found\n",
      "Processing file: P09912_blast_results.txt\n",
      "Ignoring file: P09912_blast_results.txt - No hits found\n",
      "Processing file: P01225_blast_results.txt\n",
      "Ignoring file: P01225_blast_results.txt - No hits found\n",
      "Processing file: P11802_blast_results.txt\n",
      "Processing file: Q08722_blast_results.txt\n",
      "Processing file: P01303_blast_results.txt\n",
      "Ignoring file: P01303_blast_results.txt - No hits found\n",
      "Processing file: Q15303_blast_results.txt\n",
      "Processing file: P51587_blast_results.txt\n",
      "Ignoring file: P51587_blast_results.txt - No hits found\n",
      "Processing file: Q16611_blast_results.txt\n",
      "Processing file: Q9UHF0_blast_results.txt\n",
      "Ignoring file: Q9UHF0_blast_results.txt - No hits found\n",
      "Processing file: Q01629_blast_results.txt\n",
      "Ignoring file: Q01629_blast_results.txt - No hits found\n",
      "Processing file: P0C0P6_blast_results.txt\n",
      "Ignoring file: P0C0P6_blast_results.txt - No hits found\n",
      "Processing file: P16860_blast_results.txt\n",
      "Ignoring file: P16860_blast_results.txt - No hits found\n",
      "Processing file: Q92968_blast_results.txt\n",
      "Ignoring file: Q92968_blast_results.txt - No hits found\n",
      "Processing file: P35318_blast_results.txt\n",
      "Ignoring file: P35318_blast_results.txt - No hits found\n",
      "Processing file: P47898_blast_results.txt\n",
      "Processing file: P20591_blast_results.txt\n",
      "Processing file: O43745_blast_results.txt\n",
      "Ignoring file: O43745_blast_results.txt - No hits found\n",
      "Processing file: P55089_blast_results.txt\n",
      "Ignoring file: P55089_blast_results.txt - No hits found\n",
      "Processing file: P20592_blast_results.txt\n",
      "Processing file: Q13325_blast_results.txt\n",
      "Ignoring file: Q13325_blast_results.txt - No hits found\n",
      "Processing file: P32247_blast_results.txt\n",
      "Processing file: Q9Y3Z3_blast_results.txt\n",
      "Processing file: P01160_blast_results.txt\n",
      "Ignoring file: P01160_blast_results.txt - No hits found\n",
      "Processing file: Q9H293_blast_results.txt\n",
      "Ignoring file: Q9H293_blast_results.txt - No hits found\n",
      "Processing file: P08949_blast_results.txt\n",
      "Ignoring file: P08949_blast_results.txt - No hits found\n",
      "Processing file: O95399_blast_results.txt\n",
      "Ignoring file: O95399_blast_results.txt - No hits found\n",
      "Processing file: P06396_blast_results.txt\n",
      "Processing file: P01213_blast_results.txt\n",
      "Ignoring file: P01213_blast_results.txt - No hits found\n",
      "Processing file: P01210_blast_results.txt\n",
      "Ignoring file: P01210_blast_results.txt - No hits found\n",
      "Processing file: P10082_blast_results.txt\n",
      "Ignoring file: P10082_blast_results.txt - No hits found\n",
      "Processing file: Q15726_blast_results.txt\n",
      "Ignoring file: Q15726_blast_results.txt - No hits found\n",
      "Processing file: P80303_blast_results.txt\n",
      "Ignoring file: P80303_blast_results.txt - No hits found\n",
      "Processing file: P01574_blast_results.txt\n",
      "Ignoring file: P01574_blast_results.txt - No hits found\n",
      "Processing file: Q8IUX4_blast_results.txt\n",
      "Ignoring file: Q8IUX4_blast_results.txt - No hits found\n",
      "Processing file: P40763_blast_results.txt\n",
      "Ignoring file: P40763_blast_results.txt - No hits found\n",
      "Processing file: Q00526_blast_results.txt\n",
      "Processing file: O75192_blast_results.txt\n",
      "Ignoring file: O75192_blast_results.txt - No hits found\n",
      "Processing file: P55210_blast_results.txt\n",
      "Processing file: Q02388_blast_results.txt\n",
      "Processing file: P55211_blast_results.txt\n",
      "Ignoring file: P55211_blast_results.txt - No hits found\n",
      "Processing file: A0A097EV82_blast_results.txt\n",
      "Ignoring file: A0A097EV82_blast_results.txt - No hits found\n",
      "Processing file: P13164_blast_results.txt\n",
      "Ignoring file: P13164_blast_results.txt - No hits found\n",
      "Processing file: Q96RP3_blast_results.txt\n",
      "Ignoring file: Q96RP3_blast_results.txt - No hits found\n",
      "Processing file: P23142_blast_results.txt\n",
      "Ignoring file: P23142_blast_results.txt - No hits found\n",
      "Processing file: Q00535_blast_results.txt\n",
      "Processing file: Q99653_blast_results.txt\n",
      "Ignoring file: Q99653_blast_results.txt - No hits found\n",
      "Processing file: Q01453_blast_results.txt\n",
      "Ignoring file: Q01453_blast_results.txt - No hits found\n",
      "Processing file: Q7Z2W4_blast_results.txt\n",
      "Processing file: P42226_blast_results.txt\n",
      "Ignoring file: P42226_blast_results.txt - No hits found\n",
      "Processing file: P48645_blast_results.txt\n",
      "Ignoring file: P48645_blast_results.txt - No hits found\n",
      "Processing file: P01566_blast_results.txt\n",
      "Ignoring file: P01566_blast_results.txt - No hits found\n",
      "Processing file: O15130_blast_results.txt\n",
      "Ignoring file: O15130_blast_results.txt - No hits found\n",
      "Processing file: O96011_blast_results.txt\n",
      "Ignoring file: O96011_blast_results.txt - No hits found\n",
      "Processing file: P09914_blast_results.txt\n",
      "Ignoring file: P09914_blast_results.txt - No hits found\n",
      "Processing file: Q969E3_blast_results.txt\n",
      "Ignoring file: Q969E3_blast_results.txt - No hits found\n",
      "Processing file: P55265_blast_results.txt\n",
      "Processing file: Q12965_blast_results.txt\n",
      "Processing file: P02533_blast_results.txt\n",
      "Ignoring file: P02533_blast_results.txt - No hits found\n",
      "Processing file: Q9BTW9_blast_results.txt\n",
      "Ignoring file: Q9BTW9_blast_results.txt - No hits found\n",
      "Processing file: P10645_blast_results.txt\n",
      "Processing file: Q8N729_blast_results.txt\n",
      "Ignoring file: Q8N729_blast_results.txt - No hits found\n",
      "Processing file: P07492_blast_results.txt\n",
      "Ignoring file: P07492_blast_results.txt - No hits found\n",
      "Processing file: P43490_blast_results.txt\n",
      "Processing file: Q9P0W0_blast_results.txt\n",
      "Ignoring file: Q9P0W0_blast_results.txt - No hits found\n",
      "Processing file: Q03252_blast_results.txt\n",
      "Ignoring file: Q03252_blast_results.txt - No hits found\n",
      "Processing file: P25445_blast_results.txt\n",
      "Processing file: P42224_blast_results.txt\n",
      "Ignoring file: P42224_blast_results.txt - No hits found\n",
      "Processing file: Q13519_blast_results.txt\n",
      "Ignoring file: Q13519_blast_results.txt - No hits found\n",
      "Processing file: P04637_blast_results.txt\n",
      "Ignoring file: P04637_blast_results.txt - No hits found\n",
      "Processing file: P46663_blast_results.txt\n",
      "Processing file: Q14790_blast_results.txt\n",
      "Processing file: P30990_blast_results.txt\n",
      "Ignoring file: P30990_blast_results.txt - No hits found\n",
      "Processing file: P01344_blast_results.txt\n",
      "Processing file: Q00534_blast_results.txt\n",
      "Processing file: Q9Y5Y5_blast_results.txt\n",
      "Ignoring file: Q9Y5Y5_blast_results.txt - No hits found\n",
      "Processing file: Q14765_blast_results.txt\n",
      "Processing file: Q96BS2_blast_results.txt\n",
      "Ignoring file: Q96BS2_blast_results.txt - No hits found\n",
      "Processing file: Q99988_blast_results.txt\n",
      "Ignoring file: Q99988_blast_results.txt - No hits found\n",
      "Processing file: P06400_blast_results.txt\n",
      "Ignoring file: P06400_blast_results.txt - No hits found\n",
      "Processing file: P52630_blast_results.txt\n",
      "Ignoring file: P52630_blast_results.txt - No hits found\n",
      "Processing file: P19883_blast_results.txt\n",
      "Ignoring file: P19883_blast_results.txt - No hits found\n",
      "Processing file: Q01628_blast_results.txt\n",
      "Ignoring file: Q01628_blast_results.txt - No hits found\n",
      "Processing file: Q07812_blast_results.txt\n",
      "Processing file: P01241_blast_results.txt\n",
      "Ignoring file: P01241_blast_results.txt - No hits found\n",
      "Processing file: P06493_blast_results.txt\n",
      "Processing file: Q9Y5Q6_blast_results.txt\n",
      "Ignoring file: Q9Y5Q6_blast_results.txt - No hits found\n",
      "Processing file: P42574_blast_results.txt\n",
      "Ignoring file: P42574_blast_results.txt - No hits found\n",
      "Processing file: Q8NG41_blast_results.txt\n",
      "Ignoring file: Q8NG41_blast_results.txt - No hits found\n",
      "Processing file: Q10589_blast_results.txt\n",
      "Ignoring file: Q10589_blast_results.txt - No hits found\n",
      "Processing file: P01019_blast_results.txt\n",
      "Processing file: P15502_blast_results.txt\n",
      "Ignoring file: P15502_blast_results.txt - No hits found\n",
      "Processing file: P60484_blast_results.txt\n",
      "Processing file: P19525_blast_results.txt\n",
      "Processing file: P38398_blast_results.txt\n",
      "Processing file: O00910_blast_results.txt\n",
      "Processing file: O75381_blast_results.txt\n",
      "Ignoring file: O75381_blast_results.txt - No hits found\n",
      "Included files: 41\n",
      "Ignored files: 78\n",
      "CSV generation complete.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import csv\n",
    "\n",
    "# Path to the blast outputs in fmt 5\n",
    "results_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5\"\n",
    "\n",
    "# Output CSV file path\n",
    "output_csv_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/VB_Results.csv\"\n",
    "\n",
    "# Initialize count variables\n",
    "included_files_count = 0\n",
    "ignored_files_count = 0\n",
    "\n",
    "# Write CSV header\n",
    "with open(output_csv_path, \"w\", newline=\"\") as csvfile:\n",
    "    csv_writer = csv.writer(csvfile)\n",
    "    csv_writer.writerow([\"Accession ID\", \"Hit Number\", \"Hit ID\", \"Hit Length\", \"Species\", \"Query Definition\", \"Hsp Bit Score\", \"Hsp Score\", \"Hsp E-value\", \"Identity\", \"Positives\", \"Gaps\", \"Hsp Qseq\", \"Hsp Hseq\"])\n",
    "\n",
    "# Iterate through each blast output\n",
    "for filename in os.listdir(results_directory):\n",
    "    if filename.endswith(\".txt\") and filename != \".txt\":  \n",
    "        print(f\"Processing file: {filename}\")\n",
    "        results_file_path = os.path.join(results_directory, filename)\n",
    "        \n",
    "        # Open and read the results\n",
    "        with open(results_file_path, \"r\") as results_file:\n",
    "            results_text = results_file.read()\n",
    "\n",
    "            # Check if \"0 hits found\" or \"No hits found\" is present\n",
    "            if \"0 hits found\" in results_text or \"No hits found\" in results_text:\n",
    "                print(f\"Ignoring file: {filename} - No hits found\")\n",
    "                ignored_files_count += 1\n",
    "                continue\n",
    "            \n",
    "            # Check if the file contains expected XML structure\n",
    "            if \"<Hit>\" not in results_text:\n",
    "                print(f\"Ignoring file: {filename} - Unexpected format\")\n",
    "                ignored_files_count += 1\n",
    "                continue\n",
    "            \n",
    "            # Increment included files count\n",
    "            included_files_count += 1\n",
    "            \n",
    "            # Use file names to get Accession ID's\n",
    "            accession_id = os.path.splitext(filename)[0].replace(\"_blast_results\", \"\")\n",
    "            \n",
    "            # Initialize hit number\n",
    "            hit_number = 1\n",
    "            \n",
    "            # Specify Hit ID\n",
    "            hit_ids = re.findall(r'<Hit_id>.*?\\|(.*?)\\|</Hit_id>', results_text)\n",
    "            \n",
    "            # Specify hit length\n",
    "            hit_lengths = re.findall(r'<Hit_len>(\\d+)</Hit_len>', results_text)\n",
    "            \n",
    "            # Specify species\n",
    "            species = re.findall(r'\\[(.*?)\\]', results_text)\n",
    "\n",
    "            # Specify Query Definition\n",
    "            query_definitions = re.findall(r'<Hit_def>(.*?)<', results_text)\n",
    "\n",
    "            # Specify additional Hsp information\n",
    "            hsp_bit_scores = re.findall(r'<Hsp_bit-score>(.*?)</Hsp_bit-score>', results_text)\n",
    "            hsp_scores = re.findall(r'<Hsp_score>(.*?)</Hsp_score>', results_text)\n",
    "            hsp_evals = re.findall(r'<Hsp_evalue>(.*?)</Hsp_evalue>', results_text)\n",
    "            hsp_qseqs = re.findall(r'<Hsp_qseq>(.*?)</Hsp_qseq>', results_text)\n",
    "            hsp_hseqs = re.findall(r'<Hsp_hseq>(.*?)</Hsp_hseq>', results_text)\n",
    "            hsp_identities = re.findall(r'<Hsp_identity>(.*?)</Hsp_identity>', results_text)\n",
    "            hsp_positives = re.findall(r'<Hsp_positive>(.*?)</Hsp_positive>', results_text)\n",
    "            hsp_gaps = re.findall(r'<Hsp_gaps>(.*?)</Hsp_gaps>', results_text)\n",
    "\n",
    "            # Write to CSV\n",
    "            with open(output_csv_path, \"a\", newline=\"\") as csvfile:\n",
    "                csv_writer = csv.writer(csvfile)\n",
    "                for hit_id, length, word, query_def, bit_score, score, eval, identity, positives, gaps, qseq, hseq in zip(hit_ids, hit_lengths, species, query_definitions, hsp_bit_scores, hsp_scores, hsp_evals, hsp_identities, hsp_positives, hsp_gaps, hsp_qseqs, hsp_hseqs):\n",
    "                    if query_def:  # Check if query definition is not empty\n",
    "                        csv_writer.writerow([accession_id, f\"Hit {hit_number}\", hit_id, length, word, query_def, bit_score, score, eval, identity, positives, gaps, qseq, hseq])\n",
    "                        hit_number += 1\n",
    "\n",
    "# Print count of included and ignored files\n",
    "print(f\"Included files: {included_files_count}\")\n",
    "print(f\"Ignored files: {ignored_files_count}\")\n",
    "print(\"CSV generation complete.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file: P12872_blast_results.txt\n",
      "Ignoring file: P12872_blast_results.txt - No hits found\n",
      "Processing file: P56199_blast_results.txt\n",
      "Processing file: P81605_blast_results.txt\n",
      "Ignoring file: P81605_blast_results.txt - No hits found\n",
      "Processing file: P01579_blast_results.txt\n",
      "Ignoring file: P01579_blast_results.txt - No hits found\n",
      "Processing file: P18509_blast_results.txt\n",
      "Ignoring file: P18509_blast_results.txt - No hits found\n",
      "Processing file: P01258_blast_results.txt\n",
      "Ignoring file: P01258_blast_results.txt - No hits found\n",
      "Processing file: P01570_blast_results.txt\n",
      "Ignoring file: P01570_blast_results.txt - No hits found\n",
      "Processing file: P23582_blast_results.txt\n",
      "Ignoring file: P23582_blast_results.txt - No hits found\n",
      "Processing file: P24941_blast_results.txt\n",
      "Processing file: P19838_blast_results.txt\n",
      "Processing file: O95390_blast_results.txt\n",
      "Processing file: Q9HC23_blast_results.txt\n",
      "Ignoring file: Q9HC23_blast_results.txt - No hits found\n",
      "Processing file: Q8NFJ6_blast_results.txt\n",
      "Processing file: O95786_blast_results.txt\n",
      "Processing file: P20366_blast_results.txt\n",
      "Ignoring file: P20366_blast_results.txt - No hits found\n",
      "Processing file: P01568_blast_results.txt\n",
      "Ignoring file: P01568_blast_results.txt - No hits found\n",
      "Processing file: O95760_blast_results.txt\n",
      "Ignoring file: O95760_blast_results.txt - No hits found\n",
      "Processing file: P06881_blast_results.txt\n",
      "Ignoring file: P06881_blast_results.txt - No hits found\n",
      "Processing file: Q5H8A3_blast_results.txt\n",
      "Ignoring file: Q5H8A3_blast_results.txt - No hits found\n",
      "Processing file: O43808_blast_results.txt\n",
      "Processing file: O14879_blast_results.txt\n",
      "Processing file: P01966_blast_results.txt\n",
      "Processing file: P01282_blast_results.txt\n",
      "Ignoring file: P01282_blast_results.txt - No hits found\n",
      "Processing file: P09912_blast_results.txt\n",
      "Ignoring file: P09912_blast_results.txt - No hits found\n",
      "Processing file: P01225_blast_results.txt\n",
      "Ignoring file: P01225_blast_results.txt - No hits found\n",
      "Processing file: P11802_blast_results.txt\n",
      "Processing file: P01303_blast_results.txt\n",
      "Ignoring file: P01303_blast_results.txt - No hits found\n",
      "Processing file: Q15303_blast_results.txt\n",
      "Processing file: P51587_blast_results.txt\n",
      "Ignoring file: P51587_blast_results.txt - No hits found\n",
      "Processing file: Q16611_blast_results.txt\n",
      "Processing file: Q9UHF0_blast_results.txt\n",
      "Ignoring file: Q9UHF0_blast_results.txt - No hits found\n",
      "Processing file: Q01629_blast_results.txt\n",
      "Ignoring file: Q01629_blast_results.txt - No hits found\n",
      "Processing file: P0C0P6_blast_results.txt\n",
      "Ignoring file: P0C0P6_blast_results.txt - No hits found\n",
      "Processing file: P16860_blast_results.txt\n",
      "Ignoring file: P16860_blast_results.txt - No hits found\n",
      "Processing file: Q92968_blast_results.txt\n",
      "Ignoring file: Q92968_blast_results.txt - No hits found\n",
      "Processing file: P35318_blast_results.txt\n",
      "Ignoring file: P35318_blast_results.txt - No hits found\n",
      "Processing file: P47898_blast_results.txt\n",
      "Processing file: P20591_blast_results.txt\n",
      "Processing file: O43745_blast_results.txt\n",
      "Ignoring file: O43745_blast_results.txt - No hits found\n",
      "Processing file: P55089_blast_results.txt\n",
      "Ignoring file: P55089_blast_results.txt - No hits found\n",
      "Processing file: P20592_blast_results.txt\n",
      "Processing file: Q13325_blast_results.txt\n",
      "Ignoring file: Q13325_blast_results.txt - No hits found\n",
      "Processing file: P32247_blast_results.txt\n",
      "Processing file: Q9Y3Z3_blast_results.txt\n",
      "Processing file: P01160_blast_results.txt\n",
      "Ignoring file: P01160_blast_results.txt - No hits found\n",
      "Processing file: Q9H293_blast_results.txt\n",
      "Ignoring file: Q9H293_blast_results.txt - No hits found\n",
      "Processing file: P08949_blast_results.txt\n",
      "Ignoring file: P08949_blast_results.txt - No hits found\n",
      "Processing file: O95399_blast_results.txt\n",
      "Ignoring file: O95399_blast_results.txt - No hits found\n",
      "Processing file: P06396_blast_results.txt\n",
      "Processing file: P01213_blast_results.txt\n",
      "Ignoring file: P01213_blast_results.txt - No hits found\n",
      "Processing file: P01210_blast_results.txt\n",
      "Ignoring file: P01210_blast_results.txt - No hits found\n",
      "Processing file: P10082_blast_results.txt\n",
      "Ignoring file: P10082_blast_results.txt - No hits found\n",
      "Processing file: Q15726_blast_results.txt\n",
      "Ignoring file: Q15726_blast_results.txt - No hits found\n",
      "Processing file: P80303_blast_results.txt\n",
      "Ignoring file: P80303_blast_results.txt - No hits found\n",
      "Processing file: P01574_blast_results.txt\n",
      "Ignoring file: P01574_blast_results.txt - No hits found\n",
      "Processing file: Q8IUX4_blast_results.txt\n",
      "Ignoring file: Q8IUX4_blast_results.txt - No hits found\n",
      "Processing file: P40763_blast_results.txt\n",
      "Ignoring file: P40763_blast_results.txt - No hits found\n",
      "Processing file: Q00526_blast_results.txt\n",
      "Processing file: O75192_blast_results.txt\n",
      "Ignoring file: O75192_blast_results.txt - No hits found\n",
      "Processing file: P55210_blast_results.txt\n",
      "Processing file: Q02388_blast_results.txt\n",
      "Processing file: P55211_blast_results.txt\n",
      "Ignoring file: P55211_blast_results.txt - No hits found\n",
      "Processing file: A0A097EV82_blast_results.txt\n",
      "Ignoring file: A0A097EV82_blast_results.txt - No hits found\n",
      "Processing file: P13164_blast_results.txt\n",
      "Ignoring file: P13164_blast_results.txt - No hits found\n",
      "Processing file: Q96RP3_blast_results.txt\n",
      "Ignoring file: Q96RP3_blast_results.txt - No hits found\n",
      "Processing file: P23142_blast_results.txt\n",
      "Ignoring file: P23142_blast_results.txt - No hits found\n",
      "Processing file: Q00535_blast_results.txt\n",
      "Processing file: Q99653_blast_results.txt\n",
      "Ignoring file: Q99653_blast_results.txt - No hits found\n",
      "Processing file: Q01453_blast_results.txt\n",
      "Ignoring file: Q01453_blast_results.txt - No hits found\n",
      "Processing file: Q7Z2W4_blast_results.txt\n",
      "Processing file: P42226_blast_results.txt\n",
      "Ignoring file: P42226_blast_results.txt - No hits found\n",
      "Processing file: P48645_blast_results.txt\n",
      "Ignoring file: P48645_blast_results.txt - No hits found\n",
      "Processing file: P01566_blast_results.txt\n",
      "Ignoring file: P01566_blast_results.txt - No hits found\n",
      "Processing file: O15130_blast_results.txt\n",
      "Ignoring file: O15130_blast_results.txt - No hits found\n",
      "Processing file: O96011_blast_results.txt\n",
      "Ignoring file: O96011_blast_results.txt - No hits found\n",
      "Processing file: P09914_blast_results.txt\n",
      "Ignoring file: P09914_blast_results.txt - No hits found\n",
      "Processing file: Q969E3_blast_results.txt\n",
      "Ignoring file: Q969E3_blast_results.txt - No hits found\n",
      "Processing file: P55265_blast_results.txt\n",
      "Processing file: Q12965_blast_results.txt\n",
      "Processing file: P02533_blast_results.txt\n",
      "Ignoring file: P02533_blast_results.txt - No hits found\n",
      "Processing file: Q9BTW9_blast_results.txt\n",
      "Ignoring file: Q9BTW9_blast_results.txt - No hits found\n",
      "Processing file: P10645_blast_results.txt\n",
      "Processing file: Q8N729_blast_results.txt\n",
      "Ignoring file: Q8N729_blast_results.txt - No hits found\n",
      "Processing file: P07492_blast_results.txt\n",
      "Ignoring file: P07492_blast_results.txt - No hits found\n",
      "Processing file: P43490_blast_results.txt\n",
      "Processing file: Q9P0W0_blast_results.txt\n",
      "Ignoring file: Q9P0W0_blast_results.txt - No hits found\n",
      "Processing file: Q03252_blast_results.txt\n",
      "Ignoring file: Q03252_blast_results.txt - No hits found\n",
      "Processing file: P25445_blast_results.txt\n",
      "Processing file: P42224_blast_results.txt\n",
      "Ignoring file: P42224_blast_results.txt - No hits found\n",
      "Processing file: Q13519_blast_results.txt\n",
      "Ignoring file: Q13519_blast_results.txt - No hits found\n",
      "Processing file: P04637_blast_results.txt\n",
      "Ignoring file: P04637_blast_results.txt - No hits found\n",
      "Processing file: P46663_blast_results.txt\n",
      "Processing file: Q14790_blast_results.txt\n",
      "Processing file: P30990_blast_results.txt\n",
      "Ignoring file: P30990_blast_results.txt - No hits found\n",
      "Processing file: P01344_blast_results.txt\n",
      "Processing file: Q00534_blast_results.txt\n",
      "Processing file: Q9Y5Y5_blast_results.txt\n",
      "Ignoring file: Q9Y5Y5_blast_results.txt - No hits found\n",
      "Processing file: Q14765_blast_results.txt\n",
      "Processing file: Q96BS2_blast_results.txt\n",
      "Ignoring file: Q96BS2_blast_results.txt - No hits found\n",
      "Processing file: Q99988_blast_results.txt\n",
      "Ignoring file: Q99988_blast_results.txt - No hits found\n",
      "Processing file: P06400_blast_results.txt\n",
      "Ignoring file: P06400_blast_results.txt - No hits found\n",
      "Processing file: P52630_blast_results.txt\n",
      "Ignoring file: P52630_blast_results.txt - No hits found\n",
      "Processing file: P19883_blast_results.txt\n",
      "Ignoring file: P19883_blast_results.txt - No hits found\n",
      "Processing file: Q01628_blast_results.txt\n",
      "Ignoring file: Q01628_blast_results.txt - No hits found\n",
      "Processing file: Q07812_blast_results.txt\n",
      "Processing file: P01241_blast_results.txt\n",
      "Ignoring file: P01241_blast_results.txt - No hits found\n",
      "Processing file: P06493_blast_results.txt\n",
      "Processing file: Q9Y5Q6_blast_results.txt\n",
      "Ignoring file: Q9Y5Q6_blast_results.txt - No hits found\n",
      "Processing file: P42574_blast_results.txt\n",
      "Ignoring file: P42574_blast_results.txt - No hits found\n",
      "Processing file: Q8NG41_blast_results.txt\n",
      "Ignoring file: Q8NG41_blast_results.txt - No hits found\n",
      "Processing file: Q10589_blast_results.txt\n",
      "Ignoring file: Q10589_blast_results.txt - No hits found\n",
      "Processing file: P01019_blast_results.txt\n",
      "Processing file: P15502_blast_results.txt\n",
      "Ignoring file: P15502_blast_results.txt - No hits found\n",
      "Processing file: P60484_blast_results.txt\n",
      "Processing file: P19525_blast_results.txt\n",
      "Processing file: P38398_blast_results.txt\n",
      "Processing file: O00910_blast_results.txt\n",
      "Processing file: O75381_blast_results.txt\n",
      "Ignoring file: O75381_blast_results.txt - No hits found\n",
      "Ignored files: 78\n",
      "CSV generation for no-hit queries complete.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import csv\n",
    "\n",
    "# Path to the blast outputs in fmt 5\n",
    "results_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5\"\n",
    "\n",
    "# Output CSV file path for no-hit queries\n",
    "no_hits_csv_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/No_Hits_Results.csv\"\n",
    "\n",
    "# Initialize count variables\n",
    "ignored_files_count = 0\n",
    "\n",
    "# Write CSV header for no-hit queries\n",
    "with open(no_hits_csv_path, \"w\", newline=\"\") as no_hits_file:\n",
    "    no_hits_writer = csv.writer(no_hits_file)\n",
    "    no_hits_writer.writerow([\"Filename\", \"Reason\"])\n",
    "\n",
    "# Iterate through each blast output\n",
    "for filename in os.listdir(results_directory):\n",
    "    if filename.endswith(\".txt\") and filename != \".txt\":\n",
    "        print(f\"Processing file: {filename}\")\n",
    "        results_file_path = os.path.join(results_directory, filename)\n",
    "        \n",
    "        # Open and read the results\n",
    "        with open(results_file_path, \"r\") as results_file:\n",
    "            results_text = results_file.read()\n",
    "\n",
    "            # Check if \"0 hits found\" or \"No hits found\" is present\n",
    "            if \"0 hits found\" in results_text or \"No hits found\" in results_text:\n",
    "                print(f\"Ignoring file: {filename} - No hits found\")\n",
    "                ignored_files_count += 1\n",
    "                with open(no_hits_csv_path, \"a\", newline=\"\") as no_hits_file:\n",
    "                    no_hits_writer = csv.writer(no_hits_file)\n",
    "                    no_hits_writer.writerow([filename, \"No hits found\"])\n",
    "                continue\n",
    "            \n",
    "            # Check if the file contains expected XML structure\n",
    "            if \"<Hit>\" not in results_text:\n",
    "                print(f\"Ignoring file: {filename} - Unexpected format\")\n",
    "                ignored_files_count += 1\n",
    "                with open(no_hits_csv_path, \"a\", newline=\"\") as no_hits_file:\n",
    "                    no_hits_writer = csv.writer(no_hits_file)\n",
    "                    no_hits_writer.writerow([filename, \"Unexpected format\"])\n",
    "                continue\n",
    "\n",
    "# Print count of ignored files\n",
    "print(f\"Ignored files: {ignored_files_count}\")\n",
    "print(\"CSV generation for no-hit queries complete.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process files which produced No hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moved no-hit file: P12872.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P81605.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01579.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P18509.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01258.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01570.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P23582.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9HC23.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P20366.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01568.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O95760.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P06881.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q5H8A3.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01282.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P09912.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01225.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01303.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P51587.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9UHF0.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q01629.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P0C0P6.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P16860.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q92968.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P35318.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O43745.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P55089.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q13325.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01160.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9H293.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P08949.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O95399.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01213.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01210.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P10082.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q15726.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P80303.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01574.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q8IUX4.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P40763.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O75192.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P55211.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: A0A097EV82.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P13164.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q96RP3.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P23142.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q99653.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q01453.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P42226.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P48645.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01566.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O15130.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O96011.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P09914.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q969E3.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P02533.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9BTW9.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q8N729.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P07492.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9P0W0.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q03252.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P42224.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q13519.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P04637.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P30990.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9Y5Y5.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q96BS2.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q99988.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P06400.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P52630.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P19883.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q01628.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P01241.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q9Y5Q6.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P42574.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q8NG41.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: Q10589.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: P15502.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Moved no-hit file: O75381.fasta to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Ignored (no-hit) files: 78\n",
      "No-hit files moved to the designated directory.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import shutil\n",
    "\n",
    "# Paths\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta\"\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/\"\n",
    "no_hits_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\"\n",
    "\n",
    "# Create an output directory for files which produced no hits\n",
    "os.makedirs(no_hits_directory, exist_ok=True)\n",
    "\n",
    "# Initialize count variables\n",
    "ignored_files_count = 0\n",
    "\n",
    "# Iterate through each BLAST output\n",
    "for filename in os.listdir(blast_output_directory):\n",
    "    if filename.endswith(\".txt\") and filename != \".txt\":\n",
    "        results_file_path = os.path.join(blast_output_directory, filename)\n",
    "        \n",
    "        # Open and read the results\n",
    "        with open(results_file_path, \"r\") as results_file:\n",
    "            results_text = results_file.read()\n",
    "            \n",
    "            # Check if \"0 hits found\" or \"No hits found\" is present\n",
    "            if \"0 hits found\" in results_text or \"No hits found\" in results_text:\n",
    "                ignored_files_count += 1\n",
    "                # Determine the original fasta file\n",
    "                fasta_filename = filename.replace(\"_blast_results.txt\", \".fasta\")\n",
    "                fasta_file_path = os.path.join(fasta_directory, fasta_filename)\n",
    "                \n",
    "                # Move the original fasta file to the no_hits directory\n",
    "                if os.path.exists(fasta_file_path):\n",
    "                    shutil.move(fasta_file_path, os.path.join(no_hits_directory, fasta_filename))\n",
    "                    print(f\"Moved no-hit file: {fasta_filename} to {no_hits_directory}\")\n",
    "\n",
    "print(f\"Ignored (no-hit) files: {ignored_files_count}\")\n",
    "print(\"No-hit files moved to the designated directory.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check NCBI Viral Non-Redundancies Database for hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start time (Local): 2024-06-04 12:33:52\n",
      "BLAST command executed successfully for O96011.fasta.\n",
      "End time (Local): 2024-06-04 13:00:24\n",
      "Execution time for O96011.fasta: 26 minutes and 31.48 seconds\n",
      "BLAST command executed successfully for P07492.fasta.\n",
      "End time (Local): 2024-06-04 13:31:56\n",
      "Execution time for P07492.fasta: 58 minutes and 3.42 seconds\n",
      "BLAST command executed successfully for Q15726.fasta.\n",
      "End time (Local): 2024-06-04 14:48:39\n",
      "Execution time for Q15726.fasta: 134 minutes and 46.41 seconds\n",
      "BLAST command executed successfully for P06881.fasta.\n",
      "End time (Local): 2024-06-04 15:15:10\n",
      "Execution time for P06881.fasta: 161 minutes and 17.44 seconds\n",
      "BLAST command executed successfully for P20366.fasta.\n",
      "End time (Local): 2024-06-04 15:41:41\n",
      "Execution time for P20366.fasta: 187 minutes and 49.20 seconds\n",
      "BLAST command executed successfully for P35318.fasta.\n",
      "End time (Local): 2024-06-04 16:13:14\n",
      "Execution time for P35318.fasta: 219 minutes and 21.42 seconds\n",
      "BLAST command executed successfully for P12872.fasta.\n",
      "End time (Local): 2024-06-04 17:19:54\n",
      "Execution time for P12872.fasta: 286 minutes and 1.81 seconds\n",
      "BLAST command executed successfully for O95760.fasta.\n",
      "End time (Local): 2024-06-04 18:01:30\n",
      "Execution time for O95760.fasta: 327 minutes and 37.64 seconds\n",
      "BLAST command executed successfully for O43745.fasta.\n",
      "End time (Local): 2024-06-04 19:23:14\n",
      "Execution time for O43745.fasta: 409 minutes and 21.81 seconds\n",
      "BLAST command executed successfully for Q92968.fasta.\n",
      "End time (Local): 2024-06-04 19:39:43\n",
      "Execution time for Q92968.fasta: 425 minutes and 50.51 seconds\n",
      "BLAST command executed successfully for P02533.fasta.\n",
      "End time (Local): 2024-06-04 20:01:13\n",
      "Execution time for P02533.fasta: 447 minutes and 20.74 seconds\n",
      "BLAST command executed successfully for Q01629.fasta.\n",
      "End time (Local): 2024-06-04 21:17:57\n",
      "Execution time for Q01629.fasta: 524 minutes and 5.22 seconds\n",
      "BLAST command executed successfully for P0C0P6.fasta.\n",
      "End time (Local): 2024-06-04 22:29:41\n",
      "Execution time for P0C0P6.fasta: 595 minutes and 49.38 seconds\n",
      "BLAST command executed successfully for P06400.fasta.\n",
      "End time (Local): 2024-06-04 22:32:46\n",
      "Execution time for P06400.fasta: 598 minutes and 54.28 seconds\n",
      "BLAST command executed successfully for Q03252.fasta.\n",
      "End time (Local): 2024-06-04 22:36:54\n",
      "Execution time for Q03252.fasta: 603 minutes and 2.30 seconds\n",
      "BLAST command executed successfully for Q13519.fasta.\n",
      "End time (Local): 2024-06-04 22:44:12\n",
      "Execution time for Q13519.fasta: 610 minutes and 20.08 seconds\n",
      "BLAST command executed successfully for P01574.fasta.\n",
      "End time (Local): 2024-06-04 22:56:48\n",
      "Execution time for P01574.fasta: 622 minutes and 55.65 seconds\n",
      "BLAST command executed successfully for P19883.fasta.\n",
      "End time (Local): 2024-06-04 23:43:24\n",
      "Execution time for P19883.fasta: 669 minutes and 31.99 seconds\n",
      "BLAST command executed successfully for P01282.fasta.\n",
      "End time (Local): 2024-06-04 23:52:59\n",
      "Execution time for P01282.fasta: 679 minutes and 7.18 seconds\n",
      "BLAST command executed successfully for Q9Y5Q6.fasta.\n",
      "End time (Local): 2024-06-05 00:39:36\n",
      "Execution time for Q9Y5Q6.fasta: 725 minutes and 43.69 seconds\n",
      "BLAST command executed successfully for Q8IUX4.fasta.\n",
      "End time (Local): 2024-06-05 00:43:44\n",
      "Execution time for Q8IUX4.fasta: 729 minutes and 51.88 seconds\n",
      "BLAST command executed successfully for Q99988.fasta.\n",
      "End time (Local): 2024-06-05 00:47:52\n",
      "Execution time for Q99988.fasta: 733 minutes and 59.79 seconds\n",
      "BLAST command executed successfully for P48645.fasta.\n",
      "End time (Local): 2024-06-05 00:50:57\n",
      "Execution time for P48645.fasta: 737 minutes and 4.74 seconds\n",
      "BLAST command executed successfully for Q96BS2.fasta.\n",
      "End time (Local): 2024-06-05 00:56:27\n",
      "Execution time for Q96BS2.fasta: 742 minutes and 35.34 seconds\n",
      "BLAST command executed successfully for P42226.fasta.\n",
      "End time (Local): 2024-06-05 01:28:00\n",
      "Execution time for P42226.fasta: 774 minutes and 8.07 seconds\n",
      "BLAST command failed for P09914.fasta with error: Error: NCBI C++ Exception:\n",
      "    T0 \"/home/coremake/release_build/build/PrepareRelease_Linux64-Centos_JSID_01_880026_130.14.18.128_9008__PrepareRelease_Linux64-Centos_1697736677/c++/compilers/unix/../../src/serial/rpcbase.cpp\", line 325: Error: (CRPCClientException::eFailed) ncbi::CRPCClient_Base::x_Ask() - Failed to receive reply after 3 tries (Blast4-request)\n",
      "\n",
      "\n",
      "End time (Local): 2024-06-05 01:29:36\n",
      "Execution time for P09914.fasta: 775 minutes and 44.09 seconds\n",
      "BLAST command failed for O75192.fasta with error: Error: NCBI C++ Exception:\n",
      "    T0 \"/home/coremake/release_build/build/PrepareRelease_Linux64-Centos_JSID_01_880026_130.14.18.128_9008__PrepareRelease_Linux64-Centos_1697736677/c++/compilers/unix/../../src/serial/rpcbase.cpp\", line 325: Error: (CRPCClientException::eFailed) ncbi::CRPCClient_Base::x_Ask() - Failed to receive reply after 3 tries (Blast4-request)\n",
      "\n",
      "\n",
      "End time (Local): 2024-06-05 01:31:12\n",
      "Execution time for O75192.fasta: 777 minutes and 20.09 seconds\n",
      "BLAST command executed successfully for O95399.fasta.\n",
      "End time (Local): 2024-06-05 02:18:20\n",
      "Execution time for O95399.fasta: 824 minutes and 28.16 seconds\n",
      "BLAST command failed for P08949.fasta with error: Error: NCBI C++ Exception:\n",
      "    T0 \"/home/coremake/release_build/build/PrepareRelease_Linux64-Centos_JSID_01_880026_130.14.18.128_9008__PrepareRelease_Linux64-Centos_1697736677/c++/compilers/unix/../../src/serial/rpcbase.cpp\", line 325: Error: (CRPCClientException::eFailed) ncbi::CRPCClient_Base::x_Ask() - Failed to receive reply after 3 tries (Blast4-request)\n",
      "\n",
      "\n",
      "End time (Local): 2024-06-05 02:19:56\n",
      "Execution time for P08949.fasta: 826 minutes and 4.20 seconds\n",
      "BLAST command failed for Q01453.fasta with error: Error: NCBI C++ Exception:\n",
      "    T0 \"/home/coremake/release_build/build/PrepareRelease_Linux64-Centos_JSID_01_880026_130.14.18.128_9008__PrepareRelease_Linux64-Centos_1697736677/c++/compilers/unix/../../src/serial/rpcbase.cpp\", line 325: Error: (CRPCClientException::eFailed) ncbi::CRPCClient_Base::x_Ask() - Failed to receive reply after 3 tries (Blast4-request)\n",
      "\n",
      "\n",
      "End time (Local): 2024-06-05 02:21:32\n",
      "Execution time for Q01453.fasta: 827 minutes and 40.34 seconds\n",
      "BLAST command executed successfully for Q10589.fasta.\n",
      "End time (Local): 2024-06-05 02:25:50\n",
      "Execution time for Q10589.fasta: 831 minutes and 58.10 seconds\n",
      "BLAST command executed successfully for Q969E3.fasta.\n",
      "End time (Local): 2024-06-05 02:39:18\n",
      "Execution time for Q969E3.fasta: 845 minutes and 25.72 seconds\n",
      "BLAST command executed successfully for A0A097EV82.fasta.\n",
      "End time (Local): 2024-06-05 03:01:07\n",
      "Execution time for A0A097EV82.fasta: 867 minutes and 15.39 seconds\n",
      "BLAST command executed successfully for P55089.fasta.\n",
      "End time (Local): 2024-06-05 03:17:36\n",
      "Execution time for P55089.fasta: 883 minutes and 43.89 seconds\n",
      "BLAST command executed successfully for O75381.fasta.\n",
      "End time (Local): 2024-06-05 03:24:53\n",
      "Execution time for O75381.fasta: 891 minutes and 1.28 seconds\n",
      "BLAST command executed successfully for Q9P0W0.fasta.\n",
      "End time (Local): 2024-06-05 03:37:29\n",
      "Execution time for Q9P0W0.fasta: 903 minutes and 37.14 seconds\n",
      "BLAST command executed successfully for Q9H293.fasta.\n",
      "End time (Local): 2024-06-05 03:53:58\n",
      "Execution time for Q9H293.fasta: 920 minutes and 5.55 seconds\n",
      "BLAST command executed successfully for Q9BTW9.fasta.\n",
      "End time (Local): 2024-06-05 04:10:26\n",
      "Execution time for Q9BTW9.fasta: 936 minutes and 34.16 seconds\n",
      "BLAST command executed successfully for P42224.fasta.\n",
      "End time (Local): 2024-06-05 04:23:02\n",
      "Execution time for P42224.fasta: 949 minutes and 9.86 seconds\n",
      "BLAST command executed successfully for Q9HC23.fasta.\n",
      "End time (Local): 2024-06-05 04:23:48\n",
      "Execution time for Q9HC23.fasta: 949 minutes and 56.07 seconds\n",
      "BLAST command executed successfully for P01241.fasta.\n",
      "End time (Local): 2024-06-05 04:36:25\n",
      "Execution time for P01241.fasta: 962 minutes and 32.81 seconds\n",
      "BLAST command executed successfully for P01258.fasta.\n",
      "End time (Local): 2024-06-05 04:41:55\n",
      "Execution time for P01258.fasta: 968 minutes and 3.07 seconds\n",
      "BLAST command executed successfully for Q8NG41.fasta.\n",
      "End time (Local): 2024-06-05 04:54:30\n",
      "Execution time for Q8NG41.fasta: 980 minutes and 38.29 seconds\n",
      "BLAST command executed successfully for P10082.fasta.\n",
      "End time (Local): 2024-06-05 04:58:38\n",
      "Execution time for P10082.fasta: 984 minutes and 46.00 seconds\n",
      "BLAST command executed successfully for Q96RP3.fasta.\n",
      "End time (Local): 2024-06-05 05:08:18\n",
      "Execution time for Q96RP3.fasta: 994 minutes and 25.53 seconds\n",
      "BLAST command executed successfully for P01566.fasta.\n",
      "End time (Local): 2024-06-05 05:12:27\n",
      "Execution time for P01566.fasta: 998 minutes and 34.61 seconds\n",
      "BLAST command executed successfully for P01568.fasta.\n",
      "End time (Local): 2024-06-05 05:19:45\n",
      "Execution time for P01568.fasta: 1005 minutes and 52.85 seconds\n",
      "BLAST command executed successfully for P80303.fasta.\n",
      "End time (Local): 2024-06-05 05:32:20\n",
      "Execution time for P80303.fasta: 1018 minutes and 27.96 seconds\n",
      "BLAST command executed successfully for O15130.fasta.\n",
      "End time (Local): 2024-06-05 05:39:38\n",
      "Execution time for O15130.fasta: 1025 minutes and 45.61 seconds\n",
      "BLAST command executed successfully for Q99653.fasta.\n",
      "End time (Local): 2024-06-05 05:49:13\n",
      "Execution time for Q99653.fasta: 1035 minutes and 20.54 seconds\n",
      "BLAST command executed successfully for P01570.fasta.\n",
      "End time (Local): 2024-06-05 05:58:48\n",
      "Execution time for P01570.fasta: 1044 minutes and 55.55 seconds\n",
      "BLAST command executed successfully for P30990.fasta.\n",
      "End time (Local): 2024-06-05 06:00:27\n",
      "Execution time for P30990.fasta: 1046 minutes and 34.69 seconds\n",
      "BLAST command executed successfully for P23582.fasta.\n",
      "End time (Local): 2024-06-05 06:02:04\n",
      "Execution time for P23582.fasta: 1048 minutes and 12.02 seconds\n",
      "BLAST command executed successfully for P42574.fasta.\n",
      "End time (Local): 2024-06-05 06:03:44\n",
      "Execution time for P42574.fasta: 1049 minutes and 51.94 seconds\n",
      "BLAST command executed successfully for P16860.fasta.\n",
      "End time (Local): 2024-06-05 06:06:00\n",
      "Execution time for P16860.fasta: 1052 minutes and 7.50 seconds\n",
      "BLAST command executed successfully for P01579.fasta.\n",
      "End time (Local): 2024-06-05 06:08:15\n",
      "Execution time for P01579.fasta: 1054 minutes and 22.94 seconds\n",
      "BLAST command executed successfully for Q8N729.fasta.\n",
      "End time (Local): 2024-06-05 06:15:35\n",
      "Execution time for Q8N729.fasta: 1061 minutes and 43.38 seconds\n",
      "BLAST command executed successfully for P13164.fasta.\n",
      "End time (Local): 2024-06-05 06:19:43\n",
      "Execution time for P13164.fasta: 1065 minutes and 50.96 seconds\n",
      "BLAST command executed successfully for P18509.fasta.\n",
      "End time (Local): 2024-06-05 06:41:15\n",
      "Execution time for P18509.fasta: 1087 minutes and 22.99 seconds\n",
      "BLAST command executed successfully for P01160.fasta.\n",
      "End time (Local): 2024-06-05 06:57:43\n",
      "Execution time for P01160.fasta: 1103 minutes and 51.16 seconds\n",
      "BLAST command executed successfully for P81605.fasta.\n",
      "End time (Local): 2024-06-05 07:14:12\n",
      "Execution time for P81605.fasta: 1120 minutes and 19.62 seconds\n",
      "BLAST command executed successfully for P09912.fasta.\n",
      "End time (Local): 2024-06-05 07:30:40\n",
      "Execution time for P09912.fasta: 1136 minutes and 47.75 seconds\n",
      "BLAST command executed successfully for Q13325.fasta.\n",
      "End time (Local): 2024-06-05 07:43:15\n",
      "Execution time for Q13325.fasta: 1149 minutes and 23.02 seconds\n",
      "BLAST command executed successfully for P52630.fasta.\n",
      "End time (Local): 2024-06-05 07:59:44\n",
      "Execution time for P52630.fasta: 1165 minutes and 52.30 seconds\n",
      "BLAST command executed successfully for P01210.fasta.\n",
      "End time (Local): 2024-06-05 08:16:12\n",
      "Execution time for P01210.fasta: 1182 minutes and 20.35 seconds\n",
      "BLAST command executed successfully for P40763.fasta.\n",
      "End time (Local): 2024-06-05 08:19:17\n",
      "Execution time for P40763.fasta: 1185 minutes and 24.98 seconds\n",
      "BLAST command executed successfully for P01213.fasta.\n",
      "End time (Local): 2024-06-05 08:40:46\n",
      "Execution time for P01213.fasta: 1206 minutes and 54.24 seconds\n",
      "BLAST command executed successfully for Q01628.fasta.\n",
      "End time (Local): 2024-06-05 08:57:15\n",
      "Execution time for Q01628.fasta: 1223 minutes and 22.46 seconds\n",
      "BLAST command executed successfully for P23142.fasta.\n",
      "End time (Local): 2024-06-05 09:01:22\n",
      "Execution time for P23142.fasta: 1227 minutes and 30.05 seconds\n",
      "BLAST command executed successfully for Q9UHF0.fasta.\n",
      "End time (Local): 2024-06-05 09:13:57\n",
      "Execution time for Q9UHF0.fasta: 1240 minutes and 5.40 seconds\n",
      "BLAST command executed successfully for P51587.fasta.\n",
      "End time (Local): 2024-06-05 09:23:34\n",
      "Execution time for P51587.fasta: 1249 minutes and 41.50 seconds\n",
      "BLAST command executed successfully for P15502.fasta.\n",
      "End time (Local): 2024-06-05 10:00:07\n",
      "Execution time for P15502.fasta: 1286 minutes and 14.96 seconds\n",
      "BLAST command executed successfully for P04637.fasta.\n",
      "End time (Local): 2024-06-05 10:46:43\n",
      "Execution time for P04637.fasta: 1332 minutes and 50.97 seconds\n",
      "BLAST command executed successfully for Q5H8A3.fasta.\n",
      "End time (Local): 2024-06-05 10:52:13\n",
      "Execution time for Q5H8A3.fasta: 1338 minutes and 21.22 seconds\n",
      "BLAST command executed successfully for P55211.fasta.\n",
      "End time (Local): 2024-06-05 11:33:48\n",
      "Execution time for P55211.fasta: 1379 minutes and 55.98 seconds\n",
      "BLAST command executed successfully for Q9Y5Y5.fasta.\n",
      "End time (Local): 2024-06-05 12:10:22\n",
      "Execution time for Q9Y5Y5.fasta: 1416 minutes and 29.81 seconds\n",
      "BLAST command executed successfully for P01303.fasta.\n",
      "End time (Local): 2024-06-05 13:07:00\n",
      "Execution time for P01303.fasta: 1473 minutes and 7.98 seconds\n",
      "BLAST command executed successfully for P01225.fasta.\n",
      "End time (Local): 2024-06-05 13:08:09\n",
      "Execution time for P01225.fasta: 1474 minutes and 16.47 seconds\n",
      "Total execution time: 1474 minutes and 16.47 seconds\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "\n",
    "# Paths\n",
    "no_hits_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\"\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/nr_blastp_results\"\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# BLAST parameters\n",
    "db = \"nr\"\n",
    "evalue = \"0.06\"\n",
    "remote = \"-remote\"\n",
    "entrez_query = '\"viruses[orgn]\"' \n",
    "\n",
    "# Record the start time\n",
    "start_time = time.time()\n",
    "start_time_local = time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(start_time))\n",
    "print(f\"Start time (Local): {start_time_local}\")\n",
    "\n",
    "# Iterate through the FASTA files in the no_hits directory\n",
    "for filename in os.listdir(no_hits_directory):\n",
    "    if filename.endswith(\".fasta\"):\n",
    "        fasta_file = os.path.join(no_hits_directory, filename)\n",
    "        output_file = os.path.splitext(filename)[0] + \"_blastp_results.txt\"\n",
    "        output_path = os.path.join(blast_output_directory, output_file)\n",
    "        \n",
    "        # Define the BLAST command\n",
    "        command = f\"blastp -db {db} -query {fasta_file} -out {output_path} -evalue {evalue} {remote} -entrez_query {entrez_query}\"\n",
    "        \n",
    "        # Execute the BLAST command\n",
    "        try:\n",
    "            result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "            if result.returncode == 0:\n",
    "                print(f\"BLAST command executed successfully for {filename}.\")\n",
    "            else:\n",
    "                print(f\"BLAST command failed for {filename} with return code {result.returncode}.\")\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"BLAST command failed for {filename} with error: {e.stderr.decode()}\")\n",
    "        \n",
    "        # Record the end time\n",
    "        end_time = time.time()\n",
    "        end_time_local = time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(end_time))\n",
    "\n",
    "        # Calculate the duration\n",
    "        duration = end_time - start_time\n",
    "        duration_minutes, duration_seconds = divmod(duration, 60)\n",
    "\n",
    "        print(f\"End time (Local): {end_time_local}\")\n",
    "        print(f\"Execution time for {filename}: {int(duration_minutes)} minutes and {duration_seconds:.2f} seconds\")\n",
    "\n",
    "# Calculate total run time\n",
    "total_duration = time.time() - start_time\n",
    "total_duration_minutes, total_duration_seconds = divmod(total_duration, 60)\n",
    "\n",
    "print(f\"Total execution time: {int(total_duration_minutes)} minutes and {total_duration_seconds:.2f} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process started at: 2024-07-06 18:02:17 Dublin local time\n",
      "Skipping O96011.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O96011_blastp_results.txt already exists.\n",
      "Skipping P07492.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P07492_blastp_results.txt already exists.\n",
      "Skipping Q15726.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q15726_blastp_results.txt already exists.\n",
      "Skipping P06881.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P06881_blastp_results.txt already exists.\n",
      "Skipping P20366.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P20366_blastp_results.txt already exists.\n",
      "Skipping P35318.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P35318_blastp_results.txt already exists.\n",
      "Skipping P12872.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P12872_blastp_results.txt already exists.\n",
      "Skipping O95760.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O95760_blastp_results.txt already exists.\n",
      "Skipping O43745.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O43745_blastp_results.txt already exists.\n",
      "Skipping Q92968.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q92968_blastp_results.txt already exists.\n",
      "Skipping P02533.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P02533_blastp_results.txt already exists.\n",
      "Skipping Q01629.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q01629_blastp_results.txt already exists.\n",
      "Skipping P0C0P6.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P0C0P6_blastp_results.txt already exists.\n",
      "Skipping P06400.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P06400_blastp_results.txt already exists.\n",
      "Skipping Q03252.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q03252_blastp_results.txt already exists.\n",
      "Skipping Q13519.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q13519_blastp_results.txt already exists.\n",
      "Skipping P01574.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01574_blastp_results.txt already exists.\n",
      "Skipping P19883.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P19883_blastp_results.txt already exists.\n",
      "Skipping P01282.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01282_blastp_results.txt already exists.\n",
      "Skipping Q9Y5Q6.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9Y5Q6_blastp_results.txt already exists.\n",
      "Skipping Q8IUX4.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q8IUX4_blastp_results.txt already exists.\n",
      "Skipping Q99988.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q99988_blastp_results.txt already exists.\n",
      "Skipping P48645.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P48645_blastp_results.txt already exists.\n",
      "Skipping Q96BS2.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q96BS2_blastp_results.txt already exists.\n",
      "Skipping P42226.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P42226_blastp_results.txt already exists.\n",
      "Skipping P09914.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P09914_blastp_results.txt already exists.\n",
      "Skipping concatenated_no_hits.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/concatenated_no_hits_blastp_results.txt already exists.\n",
      "Skipping O75192.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O75192_blastp_results.txt already exists.\n",
      "Skipping O95399.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O95399_blastp_results.txt already exists.\n",
      "Skipping P08949.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P08949_blastp_results.txt already exists.\n",
      "Skipping Q01453.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q01453_blastp_results.txt already exists.\n",
      "Skipping Q10589.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q10589_blastp_results.txt already exists.\n",
      "Skipping Q969E3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q969E3_blastp_results.txt already exists.\n",
      "Skipping A0A097EV82.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/A0A097EV82_blastp_results.txt already exists.\n",
      "Skipping P55089.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P55089_blastp_results.txt already exists.\n",
      "Skipping O75381.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O75381_blastp_results.txt already exists.\n",
      "Skipping Q9P0W0.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9P0W0_blastp_results.txt already exists.\n",
      "Skipping Q9H293.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9H293_blastp_results.txt already exists.\n",
      "Skipping Q9BTW9.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9BTW9_blastp_results.txt already exists.\n",
      "Skipping P42224.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P42224_blastp_results.txt already exists.\n",
      "Skipping Q9HC23.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9HC23_blastp_results.txt already exists.\n",
      "Skipping P01241.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01241_blastp_results.txt already exists.\n",
      "Skipping P01258.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01258_blastp_results.txt already exists.\n",
      "Skipping Q8NG41.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q8NG41_blastp_results.txt already exists.\n",
      "Skipping P10082.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P10082_blastp_results.txt already exists.\n",
      "Skipping Q96RP3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q96RP3_blastp_results.txt already exists.\n",
      "Skipping P01566.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01566_blastp_results.txt already exists.\n",
      "Skipping P01568.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01568_blastp_results.txt already exists.\n",
      "Skipping P80303.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P80303_blastp_results.txt already exists.\n",
      "Skipping O15130.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/O15130_blastp_results.txt already exists.\n",
      "Skipping Q99653.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q99653_blastp_results.txt already exists.\n",
      "Skipping P01570.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01570_blastp_results.txt already exists.\n",
      "Skipping P30990.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P30990_blastp_results.txt already exists.\n",
      "Skipping P23582.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P23582_blastp_results.txt already exists.\n",
      "Skipping P42574.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P42574_blastp_results.txt already exists.\n",
      "Skipping P16860.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P16860_blastp_results.txt already exists.\n",
      "Skipping P01579.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01579_blastp_results.txt already exists.\n",
      "Skipping Q8N729.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q8N729_blastp_results.txt already exists.\n",
      "Skipping P13164.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P13164_blastp_results.txt already exists.\n",
      "Skipping P18509.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P18509_blastp_results.txt already exists.\n",
      "Skipping P01160.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01160_blastp_results.txt already exists.\n",
      "Skipping P81605.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P81605_blastp_results.txt already exists.\n",
      "Skipping P09912.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P09912_blastp_results.txt already exists.\n",
      "Skipping Q13325.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q13325_blastp_results.txt already exists.\n",
      "Skipping P52630.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P52630_blastp_results.txt already exists.\n",
      "Skipping P01210.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01210_blastp_results.txt already exists.\n",
      "Skipping P40763.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P40763_blastp_results.txt already exists.\n",
      "Skipping P01213.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01213_blastp_results.txt already exists.\n",
      "Skipping Q01628.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q01628_blastp_results.txt already exists.\n",
      "Skipping P23142.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P23142_blastp_results.txt already exists.\n",
      "Skipping Q9UHF0.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9UHF0_blastp_results.txt already exists.\n",
      "Skipping P51587.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P51587_blastp_results.txt already exists.\n",
      "Skipping P15502.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P15502_blastp_results.txt already exists.\n",
      "Skipping P04637.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P04637_blastp_results.txt already exists.\n",
      "Skipping Q5H8A3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q5H8A3_blastp_results.txt already exists.\n",
      "Skipping P55211.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P55211_blastp_results.txt already exists.\n",
      "Skipping Q9Y5Y5.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Q9Y5Y5_blastp_results.txt already exists.\n",
      "Skipping P01303.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01303_blastp_results.txt already exists.\n",
      "Skipping P01225.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/P01225_blastp_results.txt already exists.\n",
      "Blastp searches completed for all fasta files. Number of files processed: 0, Number of files skipped: 79\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from datetime import datetime, timezone, timedelta\n",
    "\n",
    "# Function to convert GMT to Dublin local time\n",
    "def gmt_to_dublin(gmt_time):\n",
    "    dublin_offset = timedelta(hours=1)\n",
    "    return gmt_time + dublin_offset\n",
    "\n",
    "# Paths\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\"\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results\"\n",
    "\n",
    "# Create the blast output directory if it doesn't exist\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# Record the start time of the entire process in GMT\n",
    "start_time = datetime.now(timezone.utc)\n",
    "start_time_dublin = gmt_to_dublin(start_time)\n",
    "print(f\"Process started at: {start_time_dublin.strftime('%Y-%m-%d %H:%M:%S')} Dublin local time\")\n",
    "\n",
    "def get_accession_id(fasta_file):\n",
    "    with open(fasta_file, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>'):\n",
    "                return line.split()[0][1:]\n",
    "    return None\n",
    "\n",
    "# Initialize the file counters\n",
    "num_files_processed = 0\n",
    "num_files_skipped = 0\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):\n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        accession_id = get_accession_id(fasta_file)\n",
    "        if not accession_id:\n",
    "            print(f\"No accession ID found in {filename}, skipping...\")\n",
    "            num_files_skipped += 1\n",
    "            continue\n",
    "        \n",
    "        output_file = os.path.join(blast_output_directory, f\"{filename.split('.')[0]}_blastp_results.txt\")\n",
    "        \n",
    "        # Skip if output file already exists\n",
    "        if os.path.exists(output_file):\n",
    "            print(f\"Skipping {filename} as output file {output_file} already exists.\")\n",
    "            num_files_skipped += 1\n",
    "            continue\n",
    "        \n",
    "        # Construct the blastp command for remote search\n",
    "        blastp_command = [\n",
    "            \"blastp\",\n",
    "            \"-query\", fasta_file,\n",
    "            \"-db\", \"nr\",\n",
    "            \"-out\", output_file,\n",
    "            \"-evalue\", \"0.06\",\n",
    "            \"-entrez_query\", \"viruses[orgn]\",\n",
    "            \"-remote\"\n",
    "        ]\n",
    "        \n",
    "        # Record start time for this file\n",
    "        file_start_time = time.time()\n",
    "        \n",
    "        # Execute the blastp command\n",
    "        print(f\"Running blast search for {filename} (Accession ID: {accession_id})...\")\n",
    "        try:\n",
    "            result = subprocess.run(blastp_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "            if result.returncode == 0:\n",
    "                print(f\"BLAST command executed successfully for {filename}.\")\n",
    "            else:\n",
    "                print(f\"BLAST command failed for {filename} with return code {result.returncode}.\")\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"BLAST command failed for {filename} with error: {e.stderr.decode()}\")\n",
    "        \n",
    "        # Calculate elapsed time for this file\n",
    "        file_elapsed_time = time.time() - file_start_time\n",
    "        \n",
    "        # Print filename and elapsed time\n",
    "        print(f\"Blast search for {filename} completed in {file_elapsed_time:.2f} seconds.\")\n",
    "        \n",
    "        # Increment the file counter\n",
    "        num_files_processed += 1\n",
    "        \n",
    "        # Check if 30 minutes have elapsed since the start of the process\n",
    "        if (datetime.now(timezone.utc) - start_time).seconds >= 1800:\n",
    "            print(f\"Process has been running for 30 minutes. Current time: {gmt_to_dublin(datetime.now(timezone.utc)).strftime('%Y-%m-%d %H:%M:%S')} Dublin local time\")\n",
    "            start_time = datetime.now(timezone.utc)\n",
    "\n",
    "print(f\"Blastp searches completed for all fasta files. Number of files processed: {num_files_processed}, Number of files skipped: {num_files_skipped}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All blast_results.txt files have been concatenated into the master file: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Vir_nr_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Directory containing blast result files\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/\"\n",
    "\n",
    "# Name of the master file\n",
    "master_filename = \"Vir_nr_master_blaster.txt\"\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = os.path.join(blast_output_directory, master_filename)\n",
    "\n",
    "# Open the master file in append mode\n",
    "with open(master_filepath, \"a\") as master_file:\n",
    "    # Iterate through each file in the blast output directory\n",
    "    for filename in os.listdir(blast_output_directory):\n",
    "        # Check if the file is a blast result file\n",
    "        if filename.endswith(\"_blastp_results.txt\"):\n",
    "            # Get the full path of the blast result file\n",
    "            blast_result_filepath = os.path.join(blast_output_directory, filename)\n",
    "            # Open the blast result file and read its content\n",
    "            with open(blast_result_filepath, \"r\") as blast_result_file:\n",
    "                # Read the content of the blast result file and write it to the master file\n",
    "                master_file.write(blast_result_file.read())\n",
    "                # Write a newline character to separate the content of each blast result file\n",
    "                master_file.write(\"\\n\")\n",
    "\n",
    "print(\"All blast_results.txt files have been concatenated into the master file:\", master_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked total hits for each accession ID:\n",
      "1. Accession ID: P42574, Total Hits: 1\n",
      "Results saved to '/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Vir_nr_hit.csv'\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import csv\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Vir_nr_master_blaster.txt\"\n",
    "\n",
    "# Initialize a defaultdict to store query counts and hits\n",
    "query_hits = defaultdict(list)\n",
    "\n",
    "# Read the file and extract query IDs and hits\n",
    "try:\n",
    "    with open(master_filepath, \"r\") as master_file:\n",
    "        current_query_id = None\n",
    "        for line in master_file:\n",
    "            if \"|\" in line: \n",
    "                current_query_id = line.split(\"|\")[1].strip()  \n",
    "            elif \">\" in line:  \n",
    "                if current_query_id:\n",
    "                    query_hits[current_query_id].append(line.split(\">\")[1].strip())  \n",
    "\n",
    "    # Calculate total hits for each accession ID\n",
    "    total_hits = {query_id: len(hits) for query_id, hits in query_hits.items()}\n",
    "\n",
    "    # Rank and print total hits\n",
    "    print(\"Ranked total hits for each accession ID:\")\n",
    "    for rank, (query_id, hits) in enumerate(sorted(total_hits.items(), key=lambda x: x[1], reverse=True), start=1):\n",
    "        print(f\"{rank}. Accession ID: {query_id}, Total Hits: {hits}\")\n",
    "\n",
    "    # Write results to CSV file\n",
    "    output_csv_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/nr_blastp_results/Vir_nr_hit.csv\"\n",
    "    with open(output_csv_path, mode='w', newline='') as csv_file:\n",
    "        fieldnames = ['Query ID', 'Hits']\n",
    "        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)\n",
    "        \n",
    "        writer.writeheader()\n",
    "        for query_id, hits in query_hits.items():\n",
    "            writer.writerow({'Query ID': query_id, 'Hits': ', '.join(hits)})\n",
    "    print(f\"Results saved to '{output_csv_path}'\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: File '{master_filepath}' not found.\")\n",
    "    exit(1)\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {e}\")\n",
    "    exit(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Refseq Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process started at: 2024-07-06 18:25:39 Dublin local time\n",
      "Skipping O96011.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O96011_blastp_results.txt already exists.\n",
      "Skipping P07492.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P07492_blastp_results.txt already exists.\n",
      "Skipping Q15726.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q15726_blastp_results.txt already exists.\n",
      "Skipping P06881.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P06881_blastp_results.txt already exists.\n",
      "Skipping P20366.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P20366_blastp_results.txt already exists.\n",
      "Skipping P35318.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P35318_blastp_results.txt already exists.\n",
      "Skipping P12872.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P12872_blastp_results.txt already exists.\n",
      "Skipping O95760.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O95760_blastp_results.txt already exists.\n",
      "Skipping O43745.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O43745_blastp_results.txt already exists.\n",
      "Skipping Q92968.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q92968_blastp_results.txt already exists.\n",
      "Skipping P02533.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P02533_blastp_results.txt already exists.\n",
      "Skipping Q01629.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q01629_blastp_results.txt already exists.\n",
      "Skipping P0C0P6.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P0C0P6_blastp_results.txt already exists.\n",
      "Skipping P06400.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P06400_blastp_results.txt already exists.\n",
      "Skipping Q03252.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q03252_blastp_results.txt already exists.\n",
      "Skipping Q13519.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q13519_blastp_results.txt already exists.\n",
      "Skipping P01574.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01574_blastp_results.txt already exists.\n",
      "Skipping P19883.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P19883_blastp_results.txt already exists.\n",
      "Skipping P01282.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01282_blastp_results.txt already exists.\n",
      "Skipping Q9Y5Q6.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9Y5Q6_blastp_results.txt already exists.\n",
      "Skipping Q8IUX4.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q8IUX4_blastp_results.txt already exists.\n",
      "Skipping Q99988.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q99988_blastp_results.txt already exists.\n",
      "Skipping P48645.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P48645_blastp_results.txt already exists.\n",
      "Skipping Q96BS2.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q96BS2_blastp_results.txt already exists.\n",
      "Skipping P42226.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P42226_blastp_results.txt already exists.\n",
      "Skipping P09914.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P09914_blastp_results.txt already exists.\n",
      "Skipping concatenated_no_hits.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/concatenated_no_hits_blastp_results.txt already exists.\n",
      "Skipping O75192.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O75192_blastp_results.txt already exists.\n",
      "Skipping O95399.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O95399_blastp_results.txt already exists.\n",
      "Skipping P08949.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P08949_blastp_results.txt already exists.\n",
      "Skipping Q01453.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q01453_blastp_results.txt already exists.\n",
      "Skipping Q10589.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q10589_blastp_results.txt already exists.\n",
      "Skipping Q969E3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q969E3_blastp_results.txt already exists.\n",
      "Skipping A0A097EV82.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/A0A097EV82_blastp_results.txt already exists.\n",
      "Skipping P55089.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P55089_blastp_results.txt already exists.\n",
      "Skipping O75381.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O75381_blastp_results.txt already exists.\n",
      "Skipping Q9P0W0.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9P0W0_blastp_results.txt already exists.\n",
      "Skipping Q9H293.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9H293_blastp_results.txt already exists.\n",
      "Skipping Q9BTW9.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9BTW9_blastp_results.txt already exists.\n",
      "Skipping P42224.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P42224_blastp_results.txt already exists.\n",
      "Skipping Q9HC23.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9HC23_blastp_results.txt already exists.\n",
      "Skipping P01241.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01241_blastp_results.txt already exists.\n",
      "Skipping P01258.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01258_blastp_results.txt already exists.\n",
      "Skipping Q8NG41.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q8NG41_blastp_results.txt already exists.\n",
      "Skipping P10082.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P10082_blastp_results.txt already exists.\n",
      "Skipping Q96RP3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q96RP3_blastp_results.txt already exists.\n",
      "Skipping P01566.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01566_blastp_results.txt already exists.\n",
      "Skipping P01568.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01568_blastp_results.txt already exists.\n",
      "Skipping P80303.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P80303_blastp_results.txt already exists.\n",
      "Skipping O15130.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/O15130_blastp_results.txt already exists.\n",
      "Skipping Q99653.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q99653_blastp_results.txt already exists.\n",
      "Skipping P01570.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01570_blastp_results.txt already exists.\n",
      "Skipping P30990.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P30990_blastp_results.txt already exists.\n",
      "Skipping P23582.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P23582_blastp_results.txt already exists.\n",
      "Skipping P42574.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P42574_blastp_results.txt already exists.\n",
      "Skipping P16860.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P16860_blastp_results.txt already exists.\n",
      "Skipping P01579.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01579_blastp_results.txt already exists.\n",
      "Skipping Q8N729.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q8N729_blastp_results.txt already exists.\n",
      "Skipping P13164.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P13164_blastp_results.txt already exists.\n",
      "Skipping P18509.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P18509_blastp_results.txt already exists.\n",
      "Skipping P01160.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01160_blastp_results.txt already exists.\n",
      "Skipping P81605.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P81605_blastp_results.txt already exists.\n",
      "Skipping P09912.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P09912_blastp_results.txt already exists.\n",
      "Skipping Q13325.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q13325_blastp_results.txt already exists.\n",
      "Skipping P52630.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P52630_blastp_results.txt already exists.\n",
      "Skipping P01210.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01210_blastp_results.txt already exists.\n",
      "Skipping P40763.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P40763_blastp_results.txt already exists.\n",
      "Skipping P01213.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01213_blastp_results.txt already exists.\n",
      "Skipping Q01628.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q01628_blastp_results.txt already exists.\n",
      "Skipping P23142.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P23142_blastp_results.txt already exists.\n",
      "Skipping Q9UHF0.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9UHF0_blastp_results.txt already exists.\n",
      "Skipping P51587.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P51587_blastp_results.txt already exists.\n",
      "Skipping P15502.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P15502_blastp_results.txt already exists.\n",
      "Skipping P04637.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P04637_blastp_results.txt already exists.\n",
      "Skipping Q5H8A3.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q5H8A3_blastp_results.txt already exists.\n",
      "Skipping P55211.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P55211_blastp_results.txt already exists.\n",
      "Skipping Q9Y5Y5.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Q9Y5Y5_blastp_results.txt already exists.\n",
      "Skipping P01303.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01303_blastp_results.txt already exists.\n",
      "Skipping P01225.fasta as output file /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/P01225_blastp_results.txt already exists.\n",
      "Blastp searches completed for all fasta files. Number of files processed: 0, Number of files skipped: 79\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from datetime import datetime, timezone, timedelta\n",
    "\n",
    "# Function to convert GMT to Dublin local time\n",
    "def gmt_to_dublin(gmt_time):\n",
    "    dublin_offset = timedelta(hours=1)  \n",
    "    return gmt_time + dublin_offset\n",
    "\n",
    "# Paths\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\"\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/\"\n",
    "\n",
    "# Create the blast output directory if it doesn't exist\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# Record the start time of the entire process in GMT\n",
    "start_time = datetime.now(timezone.utc)\n",
    "start_time_dublin = gmt_to_dublin(start_time)\n",
    "print(f\"Process started at: {start_time_dublin.strftime('%Y-%m-%d %H:%M:%S')} Dublin local time\")\n",
    "\n",
    "def get_accession_id(fasta_file):\n",
    "    with open(fasta_file, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('>'):\n",
    "                return line.split()[0][1:]  \n",
    "    return None\n",
    "\n",
    "# Initialize the file counters\n",
    "num_files_processed = 0\n",
    "num_files_skipped = 0\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):\n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        accession_id = get_accession_id(fasta_file)\n",
    "        if not accession_id:\n",
    "            print(f\"No accession ID found in {filename}, skipping...\")\n",
    "            num_files_skipped += 1\n",
    "            continue\n",
    "        \n",
    "        output_file = os.path.join(blast_output_directory, f\"{filename.split('.')[0]}_blastp_results.txt\")\n",
    "        \n",
    "        # Skip if output file already exists\n",
    "        if os.path.exists(output_file):\n",
    "            print(f\"Skipping {filename} as output file {output_file} already exists.\")\n",
    "            num_files_skipped += 1\n",
    "            continue\n",
    "        \n",
    "        # Construct the blastp command for remote search\n",
    "        blastp_command = [\n",
    "            \"blastp\",\n",
    "            \"-query\", fasta_file,\n",
    "            \"-db\", \"refseq_protein\",\n",
    "            \"-out\", output_file,\n",
    "            \"-evalue\", \"0.06\",\n",
    "            \"-entrez_query\", \"viruses[orgn]\",\n",
    "            \"-remote\"\n",
    "        ]\n",
    "        \n",
    "        # Record start time for this file\n",
    "        file_start_time = time.time()\n",
    "        \n",
    "        # Execute the blastp command\n",
    "        print(f\"Running blast search for {filename} (Accession ID: {accession_id})...\")\n",
    "        try:\n",
    "            result = subprocess.run(blastp_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "            if result.returncode == 0:\n",
    "                print(f\"BLAST command executed successfully for {filename}.\")\n",
    "            else:\n",
    "                print(f\"BLAST command failed for {filename} with return code {result.returncode}.\")\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"BLAST command failed for {filename} with error: {e.stderr.decode()}\")\n",
    "        \n",
    "        # Calculate elapsed time for this file\n",
    "        file_elapsed_time = time.time() - file_start_time\n",
    "        \n",
    "        # Print filename and elapsed time\n",
    "        print(f\"Blast search for {filename} completed in {file_elapsed_time:.2f} seconds.\")\n",
    "        \n",
    "        # Increment the file counter\n",
    "        num_files_processed += 1\n",
    "        \n",
    "        # Check if 30 minutes have elapsed since the start of the process\n",
    "        if (datetime.now(timezone.utc) - start_time).seconds >= 1800:\n",
    "            print(f\"Process has been running for 30 minutes. Current time: {gmt_to_dublin(datetime.now(timezone.utc)).strftime('%Y-%m-%d %H:%M:%S')} Dublin local time\")\n",
    "            start_time = datetime.now(timezone.utc)\n",
    "\n",
    "print(f\"Blastp searches completed for all fasta files. Number of files processed: {num_files_processed}, Number of files skipped: {num_files_skipped}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mv /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/concatenated_no_hits_blastp_results.txt /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine results from Refseq Search "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All blast_results.txt files have been concatenated into the master file: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Directory containing blast result files\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/\"\n",
    "\n",
    "# Name of the master file\n",
    "master_filename = \"Vir_refseq_master_blaster.txt\"\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = os.path.join(blast_output_directory, master_filename)\n",
    "\n",
    "# Open the master file in append mode\n",
    "with open(master_filepath, \"a\") as master_file:\n",
    "    # Iterate through each file in the blast output directory\n",
    "    for filename in os.listdir(blast_output_directory):\n",
    "        # Check if the file is a blast result file\n",
    "        if filename.endswith(\"_blastp_results.txt\"):\n",
    "            # Get the full path of the blast result file\n",
    "            blast_result_filepath = os.path.join(blast_output_directory, filename)\n",
    "            # Open the blast result file and read its content\n",
    "            with open(blast_result_filepath, \"r\") as blast_result_file:\n",
    "                # Read the content of the blast result file and write it to the master file\n",
    "                master_file.write(blast_result_file.read())\n",
    "                # Write a newline character to separate the content of each blast result file\n",
    "                master_file.write(\"\\n\")\n",
    "\n",
    "print(\"All blast_results.txt files have been concatenated into the master file:\", master_filepath)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked total hits for each accession ID:\n",
      "1. Accession ID: P02533, Total Hits: 24\n",
      "2. Accession ID: Q03252, Total Hits: 20\n",
      "3. Accession ID: P42574, Total Hits: 12\n",
      "4. Accession ID: Q99988, Total Hits: 4\n",
      "5. Accession ID: P15502, Total Hits: 3\n",
      "6. Accession ID: P01241, Total Hits: 3\n",
      "7. Accession ID: Q96BS2, Total Hits: 3\n",
      "8. Accession ID: P52630, Total Hits: 3\n",
      "9. Accession ID: Q10589, Total Hits: 3\n",
      "10. Accession ID: P10082, Total Hits: 3\n",
      "11. Accession ID: P01570, Total Hits: 3\n",
      "Results saved to '/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_hit.csv'\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import csv\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_master_blaster.txt\"\n",
    "\n",
    "# Initialize a defaultdict to store query counts and hits\n",
    "query_hits = defaultdict(list)\n",
    "\n",
    "# Read the file and extract query IDs and hits\n",
    "try:\n",
    "    with open(master_filepath, \"r\") as master_file:\n",
    "        current_query_id = None\n",
    "        for line in master_file:\n",
    "            if \"|\" in line: \n",
    "                current_query_id = line.split(\"|\")[1].strip()  \n",
    "            elif \">\" in line:  \n",
    "                if current_query_id:\n",
    "                    query_hits[current_query_id].append(line.split(\">\")[1].strip())  \n",
    "\n",
    "    # Calculate total hits for each accession ID\n",
    "    total_hits = {query_id: len(hits) for query_id, hits in query_hits.items()}\n",
    "\n",
    "    # Rank and print total hits\n",
    "    print(\"Ranked total hits for each accession ID:\")\n",
    "    for rank, (query_id, hits) in enumerate(sorted(total_hits.items(), key=lambda x: x[1], reverse=True), start=1):\n",
    "        print(f\"{rank}. Accession ID: {query_id}, Total Hits: {hits}\")\n",
    "\n",
    "    # Write results to CSV file\n",
    "    output_csv_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_hit.csv\"\n",
    "    with open(output_csv_path, mode='w', newline='') as csv_file:\n",
    "        fieldnames = ['Query ID', 'Hits']\n",
    "        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)\n",
    "        \n",
    "        writer.writeheader()\n",
    "        for query_id, hits in query_hits.items():\n",
    "            writer.writerow({'Query ID': query_id, 'Hits': ', '.join(hits)})\n",
    "    print(f\"Results saved to '{output_csv_path}'\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: File '{master_filepath}' not found.\")\n",
    "    exit(1)\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {e}\")\n",
    "    exit(1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accession IDs and hit names saved to '/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_hit.csv'\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "# Path to the input text file\n",
    "input_file_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_master_blaster.txt\"\n",
    "\n",
    "# Path to the output CSV file\n",
    "output_csv_path = '/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/Vir_refseq_hit.csv'\n",
    "\n",
    "# Initialize lists to store accession IDs and hit names\n",
    "accession_ids = []\n",
    "hit_names = []\n",
    "\n",
    "# Read the input file and extract accession IDs and hit names\n",
    "try:\n",
    "    with open(input_file_path, \"r\") as input_file:\n",
    "        current_accession_id = None\n",
    "        for line in input_file:\n",
    "            if \"|\" in line:\n",
    "                current_accession_id = line.split(\"|\")[1].strip()\n",
    "            elif line.startswith(\">\"):\n",
    "                if current_accession_id:\n",
    "                    accession_ids.append(current_accession_id)\n",
    "                    hit_name = line.strip()[1:] \n",
    "                    hit_names.append(hit_name)\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: File '{input_file_path}' not found.\")\n",
    "    exit(1)\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {e}\")\n",
    "    exit(1)\n",
    "\n",
    "# Write accession IDs and hit names to CSV file\n",
    "try:\n",
    "    with open(output_csv_path, mode='w', newline='') as csv_file:\n",
    "        writer = csv.writer(csv_file)\n",
    "        writer.writerow([\"Accession ID\", \"Hit Name\"])  \n",
    "        for accession_id, hit_name in zip(accession_ids, hit_names):\n",
    "            writer.writerow([accession_id, hit_name])\n",
    "    print(f\"Accession IDs and hit names saved to '{output_csv_path}'\")\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred while writing to CSV: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSV file created successfully\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import csv\n",
    "\n",
    "# Source directory containing blast result files with hits\n",
    "source_dir =  \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/\"\n",
    "\n",
    "# Output CSV file path\n",
    "output_csv = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/refseq_blastp_results/comp_Vir_refseq_data.csv\"\n",
    "\n",
    "# Function to extract query ID from file name\n",
    "def extract_query_id(file_name):\n",
    "    return file_name.replace(\"_blastp_results.txt\", \"\")\n",
    "\n",
    "# Parse out required data from blast reaults\n",
    "def parse_blast_results(file_path):\n",
    "    hits_data = []\n",
    "    with open(file_path, 'r') as f:\n",
    "        query_id = extract_query_id(os.path.basename(file_path))\n",
    "        hit_number = 0 \n",
    "        hit_id = None \n",
    "        species_id = None \n",
    "        seq_length = None  \n",
    "        evalue = None \n",
    "        score = None  \n",
    "        identities = None  \n",
    "        identities_percent = None\n",
    "        positives = None \n",
    "        positives_percent = None  \n",
    "        gaps = None  \n",
    "        gaps_percent = None  \n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                hit_number += 1 \n",
    "                hit_id = line.split(\">\")[1].split(\" \")[0]\n",
    "                species_split = line.split(\"[\")\n",
    "                if len(species_split) > 1:\n",
    "                    species_id = species_split[1].split(\"]\")[0]\n",
    "                else:\n",
    "                    species_id = None\n",
    "               \n",
    "\n",
    "            elif hit_id and \"Length=\" in line:\n",
    "                seq_length = int(line.split(\"=\")[-1].strip())\n",
    "            elif hit_id and \"Expect =\" in line:\n",
    "                evalue = line.split(\"Expect = \")[1].split(\",\")[0].strip()\n",
    "                score = line.split(\"Score = \")[1].split(\" bits\")[0].strip()\n",
    "            elif hit_id and \"Identities =\" in line:\n",
    "                identities_str = line.split(\"Identities = \")[1].split(\",\")[0].strip()\n",
    "                identities = identities_str.split(\" \")[0]\n",
    "                identities_fraction = identities.split(\"/\")\n",
    "                identities_percent = (int(identities_fraction[0]) / int(identities_fraction[1])) * 100\n",
    "                positives_str = line.split(\"Positives = \")[1].split(\",\")[0].strip()\n",
    "                positives = positives_str.split(\" \")[0]\n",
    "                positives_fraction = positives.split(\"/\")\n",
    "                positives_percent = (int(positives_fraction[0]) / int(positives_fraction[1])) * 100\n",
    "                gaps_str = line.split(\"Gaps = \")[1].split(\" \")[0].strip()\n",
    "                gaps = gaps_str\n",
    "                gaps_fraction = gaps.split(\"/\")\n",
    "                gaps_percent = (int(gaps_fraction[0]) / int(gaps_fraction[1])) * 100\n",
    "                # Append all values to hits_data\n",
    "                hits_data.append([\n",
    "                    query_id, hit_id, hit_number, seq_length, species_id, evalue, score, \n",
    "                    identities, f\"{identities_percent:.2f}%\", positives, f\"{positives_percent:.2f}%\", gaps, f\"{gaps_percent:.2f}%\"\n",
    "                ])\n",
    "    return hits_data\n",
    "\n",
    "# List to store parsed data\n",
    "parsed_data = []\n",
    "\n",
    "# Loop through all files in the source directory\n",
    "for file_name in os.listdir(source_dir):\n",
    "    if file_name.endswith(\"_blastp_results.txt\"): \n",
    "        file_path = os.path.join(source_dir, file_name)\n",
    "        parsed_data.extend(parse_blast_results(file_path))\n",
    "\n",
    "# Write parsed data to CSV file\n",
    "with open(output_csv, 'w', newline='') as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    # Write header\n",
    "    writer.writerow([\n",
    "        'Query ID', 'Hit ID', 'Hit Number', 'Sequence Length', 'Species ID', 'Evalue', 'Score', \n",
    "        'Identities', 'Identities (%)', 'Positives', 'Positives (%)', 'Gaps', 'Gaps (%)'\n",
    "    ])\n",
    "    # Write parsed data\n",
    "    writer.writerows(parsed_data)\n",
    "\n",
    "print(\"CSV file created successfully\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Move Fasta's back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished moving 0 fasta files.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "\n",
    "# Path to the directory containing the fasta files\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta/\"\n",
    "\n",
    "# Destination directory to move the fasta files\n",
    "destination_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/matching_fasta_files/\"\n",
    "\n",
    "# Create the destination directory if it doesn't exist\n",
    "os.makedirs(destination_directory, exist_ok=True)\n",
    "\n",
    "# Counter for the number of files moved\n",
    "files_moved_count = 0\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for fasta_filename in os.listdir(fasta_directory):\n",
    "    # Check if the file is a fasta file\n",
    "    if fasta_filename.endswith(\".fasta\"):\n",
    "        # Extract accession ID from the filename\n",
    "        accession_id = fasta_filename.split(\"_\")[0]\n",
    "        \n",
    "        # Check if the accession ID is in the blast results\n",
    "        if accession_id in query_hits:\n",
    "            # Path to the fasta file\n",
    "            fasta_filepath = os.path.join(fasta_directory, fasta_filename)\n",
    "            \n",
    "            # Move the fasta file to the destination directory\n",
    "            shutil.move(fasta_filepath, destination_directory)\n",
    "            \n",
    "            # Increment the count of files moved\n",
    "            files_moved_count += 1\n",
    "            \n",
    "            print(f\"Moved {fasta_filename} to {destination_directory}\")\n",
    "\n",
    "print(f\"Finished moving {files_moved_count} fasta files.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory found: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Directory found: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/nr_blastp_results\n",
      "BLAST executable found and is executable: /home/osheakes/Research_Project_MMM/Fasta/ncbi-blast-2.15.0+/bin/tblastn_vdb\n",
      "Found 79 .fasta file(s) in: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\n",
      "Environment and path checks completed successfully.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Paths\n",
    "no_hits_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/no_hits\"\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/nr_blastp_results\"\n",
    "blast_executable = \"/home/osheakes/Research_Project_MMM/Fasta/ncbi-blast-2.15.0+/bin/tblastn_vdb\"\n",
    "db = \"viruses_wgs\"\n",
    "\n",
    "# Check if the no_hits directory exists\n",
    "if not os.path.isdir(no_hits_directory):\n",
    "    raise FileNotFoundError(f\"Directory not found: {no_hits_directory}\")\n",
    "else:\n",
    "    print(f\"Directory found: {no_hits_directory}\")\n",
    "\n",
    "# Check if the blast_output directory exists (if not, create it)\n",
    "if not os.path.isdir(blast_output_directory):\n",
    "    os.makedirs(blast_output_directory)\n",
    "    print(f\"Directory created: {blast_output_directory}\")\n",
    "else:\n",
    "    print(f\"Directory found: {blast_output_directory}\")\n",
    "\n",
    "# Check if the BLAST executable exists and is executable\n",
    "if not os.path.isfile(blast_executable):\n",
    "    raise FileNotFoundError(f\"BLAST executable not found: {blast_executable}\")\n",
    "elif not os.access(blast_executable, os.X_OK):\n",
    "    raise PermissionError(f\"BLAST executable is not executable: {blast_executable}\")\n",
    "else:\n",
    "    print(f\"BLAST executable found and is executable: {blast_executable}\")\n",
    "\n",
    "# Check if there are any .fasta files in the no_hits directory\n",
    "fasta_files = [f for f in os.listdir(no_hits_directory) if f.endswith(\".fasta\")]\n",
    "if not fasta_files:\n",
    "    raise FileNotFoundError(f\"No .fasta files found in: {no_hits_directory}\")\n",
    "else:\n",
    "    print(f\"Found {len(fasta_files)} .fasta file(s) in: {no_hits_directory}\")\n",
    "\n",
    "print(\"Environment and path checks completed successfully.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.5 Creation of Master Blast file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All blast_results.txt files have been concatenated into the master file: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/viral_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Directory containing blast result files\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts\"\n",
    "\n",
    "# Name of the master file\n",
    "master_filename = \"viral_master_blaster.txt\"\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = os.path.join(blast_output_directory, master_filename)\n",
    "\n",
    "# Open the master file in append mode\n",
    "with open(master_filepath, \"a\") as master_file:\n",
    "    # Iterate through each file in the blast output directory\n",
    "    for filename in os.listdir(blast_output_directory):\n",
    "        # Check if the file is a blast result file\n",
    "        if filename.endswith(\"_blast_results.txt\"):\n",
    "            # Get the full path of the blast result file\n",
    "            blast_result_filepath = os.path.join(blast_output_directory, filename)\n",
    "            # Open the blast result file and read its content\n",
    "            with open(blast_result_filepath, \"r\") as blast_result_file:\n",
    "                # Read the content of the blast result file and write it to the master file\n",
    "                master_file.write(blast_result_file.read())\n",
    "                # Write a newline character to separate the content of each blast result file\n",
    "                master_file.write(\"\\n\")\n",
    "\n",
    "print(\"All blast_results.txt files have been concatenated into the master file:\", master_filepath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "616\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Load the data from the CSV file\n",
    "data = pd.read_csv(\"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/fmt5/VB_Results.csv\")\n",
    "\n",
    "# Define search\n",
    "pattern = r'\\bhuman\\b'\n",
    "\n",
    "# Initialize count\n",
    "human_count = 0\n",
    "\n",
    "# Iterate over each row\n",
    "for index, row in data.iterrows():\n",
    "    if re.search(pattern, str(row['Species']), flags=re.IGNORECASE):\n",
    "        human_count += 1\n",
    "\n",
    "# Print the count\n",
    "print(human_count)\n",
    "\n",
    "# Filter rows\n",
    "human_entries = data[data['Species'].str.contains('Human', case=False)]\n",
    "\n",
    "# Save to CSV file\n",
    "human_entries.to_csv(\"human_entries.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.6 Create Counter to identify which query had the most hits across the Viral database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked list of queries based on the number of hits:\n",
      "1. Query ID: sp|Q02388|CO7A1_HUMAN, Hits: 14725\n",
      "2. Query ID: sp|P19838|NFKB1_HUMAN, Hits: 1757\n",
      "3. Query ID: #, Hits: 1499\n",
      "4. Query ID: sp|Q15303|ERBB4_HUMAN, Hits: 1487\n",
      "5. Query ID: sp|P19525|E2AK2_HUMAN, Hits: 1435\n",
      "6. Query ID: sp|Q00534|CDK6_HUMAN, Hits: 1165\n",
      "7. Query ID: sp|P06493|CDK1_HUMAN, Hits: 1127\n",
      "8. Query ID: sp|Q00526|CDK3_HUMAN, Hits: 1121\n",
      "9. Query ID: sp|P24941|CDK2_HUMAN, Hits: 1063\n",
      "10. Query ID: sp|P11802|CDK4_HUMAN, Hits: 1019\n",
      "11. Query ID: sp|Q00535|CDK5_HUMAN, Hits: 1011\n",
      "12. Query ID: sp|Q8NFJ6|PKR2_HUMAN, Hits: 1001\n",
      "13. Query ID: sp|P32247|BRS3_HUMAN, Hits: 1001\n",
      "14. Query ID: sp|P43490|NAMPT_HUMAN, Hits: 1001\n",
      "15. Query ID: sp|P46663|BKRB1_HUMAN, Hits: 1001\n",
      "16. Query ID: sp|P01019|ANGT_HUMAN, Hits: 1001\n",
      "17. Query ID: sp|P55265|DSRAD_HUMAN, Hits: 919\n",
      "18. Query ID: sp|Q9Y3Z3|SAMH1_HUMAN, Hits: 231\n",
      "19. Query ID: sp|P47898|5HT5A_HUMAN, Hits: 141\n",
      "20. Query ID: sp|P25445|TNR6_HUMAN, Hits: 141\n",
      "21. Query ID: sp|P20591|MX1_HUMAN, Hits: 63\n",
      "22. Query ID: sp|P20592|MX2_HUMAN, Hits: 61\n",
      "23. Query ID: sp|O95786|RIGI_HUMAN, Hits: 37\n",
      "24. Query ID: sp|P38398|BRCA1_HUMAN, Hits: 37\n",
      "25. Query ID: sp|O43808|PM34_HUMAN, Hits: 33\n",
      "26. Query ID: sp|P01344|IGF2_HUMAN, Hits: 29\n",
      "27. Query ID: sp|P55210|CASP7_HUMAN, Hits: 21\n",
      "28. Query ID: sp|Q07812|BAX_HUMAN, Hits: 21\n",
      "29. Query ID: sp|Q16611|BAK_HUMAN, Hits: 13\n",
      "30. Query ID: sp|P06396|GELS_HUMAN, Hits: 13\n",
      "31. Query ID: sp|O95390|GDF11_HUMAN, Hits: 9\n",
      "32. Query ID: sp|O14879|IFIT3_HUMAN, Hits: 9\n",
      "33. Query ID: sp|P60484|PTEN_HUMAN, Hits: 9\n",
      "34. Query ID: sp|Q14790|CASP8_HUMAN, Hits: 7\n",
      "35. Query ID: sp|P42574|CASP3_HUMAN, Hits: 7\n",
      "36. Query ID: sp|P56199|ITA1_HUMAN, Hits: 5\n",
      "37. Query ID: sp|Q7Z2W4|ZCCHV_HUMAN, Hits: 5\n",
      "38. Query ID: sp|Q12965|MYO1E_HUMAN, Hits: 5\n",
      "39. Query ID: sp|P10645|CMGA_HUMAN, Hits: 5\n",
      "40. Query ID: sp|P01966|HBA_BOVIN, Hits: 3\n",
      "41. Query ID: sp|Q14765|STAT4_HUMAN, Hits: 3\n",
      "42. Query ID: sp|O00910|STATA_DICDI, Hits: 3\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/viral_master_blaster.txt\"\n",
    "\n",
    "# Read the contents of the master file\n",
    "with open(master_filepath, \"r\") as master_file:\n",
    "    # Read all lines from the file\n",
    "    lines = master_file.readlines()\n",
    "\n",
    "# Extract query sequence IDs from each line\n",
    "query_ids = [line.split()[0] for line in lines if line.strip()]\n",
    "\n",
    "# Count the occurrences of each query ID\n",
    "query_counts = Counter(query_ids)\n",
    "\n",
    "# Sort the queries based on the number of hits (descending order)\n",
    "ranked_queries = sorted(query_counts.items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "# Print the ranked list of queries\n",
    "print(\"Ranked list of queries based on the number of hits:\")\n",
    "for rank, (query, count) in enumerate(ranked_queries, start=1):\n",
    "    print(f\"{rank}. Query ID: {query}, Hits: {count}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.7 Cleaning up of dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out lines with query ID \"#\"\n",
    "filtered_queries = [(query, count) for query, count in ranked_queries if query != \"#\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.8 Create a Ranked table of the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked list of queries (excluding '#') saved to: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Hits_Tab.csv\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "# Path to the output CSV file with the desired file name\n",
    "output_csv_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Hits_Tab.csv\"\n",
    "\n",
    "# Write the filtered list of queries to a CSV file\n",
    "with open(output_csv_path, \"w\", newline=\"\") as csv_file:\n",
    "    # Create a CSV writer object\n",
    "    csv_writer = csv.writer(csv_file)\n",
    "    \n",
    "    # Write the header row without the \"Rank\" column\n",
    "    csv_writer.writerow([\"Query ID\", \"Hits\"])\n",
    "    \n",
    "    # Write each ranked query to the CSV file without the rank\n",
    "    for query, count in filtered_queries:\n",
    "        csv_writer.writerow([query, count])\n",
    "\n",
    "print(\"Ranked list of queries (excluding '#') saved to:\", output_csv_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.9 Create a filtered Viral Masterfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered results saved to /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/filtered_viral_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "# Read query IDs from Ranked_Hits_Tab.csv\n",
    "query_ids_from_csv = set()\n",
    "with open(\"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Hits_Tab.csv\", \"r\") as csv_file:\n",
    "    next(csv_file)  # Skip header row\n",
    "    for line in csv_file:\n",
    "        query_id = line.strip().split(\",\")[0]\n",
    "        query_ids_from_csv.add(query_id)\n",
    "\n",
    "# Create a new .txt file to store filtered results\n",
    "output_txt_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/filtered_viral_master_blaster.txt\"\n",
    "with open(\"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/viral_master_blaster.txt\", \"r\") as input_file, open(output_txt_path, \"w\") as output_file:\n",
    "    for line in input_file:\n",
    "        query_id = line.strip().split(\"\\t\")[0]\n",
    "        if query_id in query_ids_from_csv:\n",
    "            output_file.write(line)\n",
    "\n",
    "print(f\"Filtered results saved to {output_txt_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.10 Create a Ranked Table for Sequence Homology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked hits saved to: /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Homology_Hits.csv\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "# Read the lines from the file\n",
    "with open(\"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/filtered_viral_master_blaster.txt\", \"r\") as input_file:\n",
    "    lines = input_file.readlines()\n",
    "\n",
    "# Sort the lines based on the third column (sequence homology score) from highest to lowest\n",
    "sorted_lines = sorted(lines, key=lambda x: float(x.strip().split(\"\\t\")[2]), reverse=True)\n",
    "\n",
    "# Write the sorted lines to a CSV file with the header row\n",
    "output_file_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Homology_Hits.csv\"\n",
    "with open(output_file_path, \"w\", newline=\"\") as output_file:\n",
    "    csv_writer = csv.writer(output_file)\n",
    "    \n",
    "    # Write the header row\n",
    "    csv_writer.writerow([\"query acc.ver\", \"subject acc.ver\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\", \"s. start\", \"s. end\", \"evalue\", \"bit score\"])\n",
    "    \n",
    "    # Write the sorted lines to the CSV file\n",
    "    for line in sorted_lines:\n",
    "        csv_writer.writerow(line.strip().split(\"\\t\"))\n",
    "\n",
    "print(f\"Ranked hits saved to: {output_file_path}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.11 Retrieve Sequencing Data from Queries which Returned Hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir  /home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/processed_viral_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Accession ID'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/opt/jupyterhub/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2897\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2898\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2899\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Accession ID'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-324480650910>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# Count the unique Accession IDs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0munique_accession_ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Accession ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnunique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# Print the number of unique Accession IDs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/jupyterhub/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2904\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2905\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2906\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2907\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2908\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/jupyterhub/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2898\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2899\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2900\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2901\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2902\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Accession ID'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Path to the file\n",
    "file_path = \"/home/osheakes/Research_Project_MMM/Fasta/Viral_Blasts/Ranked_Homology_Hits.csv\"\n",
    "\n",
    "# Read the file into a DataFrame\n",
    "df = pd.read_csv(file_path, delimiter=\"\\t\")  # Adjust delimiter if necessary\n",
    "\n",
    "# Count the unique Accession IDs\n",
    "unique_accession_ids = df['Accession ID'].nunique()\n",
    "\n",
    "# Print the number of unique Accession IDs\n",
    "print(f\"Number of unique Accession IDs: {unique_accession_ids}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2.0 Bacterial Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Bacterial_Blasts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!blastp -db nr -query P01225.fasta -out P01225_TEST.txt -evalue 0.06 -remote -entrez_query \"bacteria[orgn]\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping O96011.fasta as output file already exists.\n",
      "Skipping Q07812.fasta as output file already exists.\n",
      "Skipping Q8NFJ6.fasta as output file already exists.\n",
      "Skipping P07492.fasta as output file already exists.\n",
      "Skipping Q15303.fasta as output file already exists.\n",
      "Skipping Q15726.fasta as output file already exists.\n",
      "Skipping O00910.fasta as output file already exists.\n",
      "Skipping P38398.fasta as output file already exists.\n",
      "Skipping P06881.fasta as output file already exists.\n",
      "Skipping P46663.fasta as output file already exists.\n",
      "Skipping P32247.fasta as output file already exists.\n",
      "Skipping P60484.fasta as output file already exists.\n",
      "Skipping P20366.fasta as output file already exists.\n",
      "Skipping P35318.fasta as output file already exists.\n",
      "Skipping P12872.fasta as output file already exists.\n",
      "Skipping P10645.fasta as output file already exists.\n",
      "Skipping O95760.fasta as output file already exists.\n",
      "Skipping P01344.fasta as output file already exists.\n",
      "Skipping O14879.fasta as output file already exists.\n",
      "Skipping Q08722.fasta as output file already exists.\n",
      "Skipping O43745.fasta as output file already exists.\n",
      "Skipping Q92968.fasta as output file already exists.\n",
      "Skipping O95390.fasta as output file already exists.\n",
      "Skipping P02533.fasta as output file already exists.\n",
      "Skipping Q01629.fasta as output file already exists.\n",
      "Skipping P0C0P6.fasta as output file already exists.\n",
      "Skipping Q00534.fasta as output file already exists.\n",
      "Skipping P20592.fasta as output file already exists.\n",
      "Skipping Q13519.fasta as output file already exists.\n",
      "Skipping P01282.fasta as output file already exists.\n",
      "Skipping Q8IUX4.fasta as output file already exists.\n",
      "Skipping Q99988.fasta as output file already exists.\n",
      "Skipping P48645.fasta as output file already exists.\n",
      "Skipping O75192.fasta as output file already exists.\n",
      "Skipping O95399.fasta as output file already exists.\n",
      "Skipping P08949.fasta as output file already exists.\n",
      "Skipping Q969E3.fasta as output file already exists.\n",
      "Skipping A0A097EV82.fasta as output file already exists.\n",
      "Skipping P55089.fasta as output file already exists.\n",
      "Skipping Q9P0W0.fasta as output file already exists.\n",
      "Skipping Q9H293.fasta as output file already exists.\n",
      "Skipping Q9HC23.fasta as output file already exists.\n",
      "Skipping P01258.fasta as output file already exists.\n",
      "Skipping Q8NG41.fasta as output file already exists.\n",
      "Skipping P10082.fasta as output file already exists.\n",
      "Skipping Q96RP3.fasta as output file already exists.\n",
      "Skipping P01566.fasta as output file already exists.\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Path to the directory containing fasta files\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta\"\n",
    "\n",
    "# Directory to store blast results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Bacterial_Blasts\"\n",
    "\n",
    "# Create the blast output directory if it doesn't exist\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):  \n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        output_file = os.path.join(blast_output_directory, f\"{os.path.splitext(filename)[0]}_blastp_results.txt\")\n",
    "        \n",
    "        # Skip if output file already exists\n",
    "        if os.path.exists(output_file):\n",
    "            print(f\"Skipping {filename} as output file already exists.\")\n",
    "            continue\n",
    "        \n",
    "        # Construct the blastp command\n",
    "        blastp_command = [\n",
    "            \"blastp\",\n",
    "            \"-db\",\n",
    "            \"nr\",\n",
    "            \"-query\",\n",
    "            fasta_file,\n",
    "            \"-out\",\n",
    "            output_file,\n",
    "            \"-evalue\",\n",
    "            \"0.06\",\n",
    "            \"-remote\",\n",
    "            \"-entrez_query\",\n",
    "            \"bacteria[orgn]\"\n",
    "        ]\n",
    "        \n",
    "        # Execute the blastp command\n",
    "        subprocess.run(blastp_command)\n",
    "        \n",
    "\n",
    "print(\"Blastp searches completed for all fasta files.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3.0 Archaeal Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Archaeal_Blasts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1 Run Archaeal Blast "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process started at: 2024-05-02 20:59:15 GMT\n",
      "Running blast search for O96011.fasta...\n",
      "Blast search for O96011.fasta completed in 18161.19 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 02:01:56 GMT\n",
      "Running blast search for Q07812.fasta...\n",
      "Blast search for Q07812.fasta completed in 4905.02 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 03:23:41 GMT\n",
      "Running blast search for Q8NFJ6.fasta...\n",
      "Blast search for Q8NFJ6.fasta completed in 7622.79 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 05:30:44 GMT\n",
      "Running blast search for P07492.fasta...\n",
      "Blast search for P07492.fasta completed in 7013.17 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 07:27:37 GMT\n",
      "Running blast search for Q15303.fasta...\n",
      "Blast search for Q15303.fasta completed in 7326.64 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 09:29:44 GMT\n",
      "Running blast search for Q15726.fasta...\n",
      "Blast search for Q15726.fasta completed in 6410.34 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 11:16:34 GMT\n",
      "Running blast search for O00910.fasta...\n",
      "Blast search for O00910.fasta completed in 7918.89 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 13:28:33 GMT\n",
      "Running blast search for P38398.fasta...\n",
      "Blast search for P38398.fasta completed in 20881.04 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 19:16:34 GMT\n",
      "Running blast search for P06881.fasta...\n",
      "Blast search for P06881.fasta completed in 15151.98 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-03 23:29:06 GMT\n",
      "Running blast search for P46663.fasta...\n",
      "Blast search for P46663.fasta completed in 9424.94 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 02:06:11 GMT\n",
      "Running blast search for P32247.fasta...\n",
      "Blast search for P32247.fasta completed in 12660.05 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 05:37:11 GMT\n",
      "Running blast search for P60484.fasta...\n",
      "Blast search for P60484.fasta completed in 107.31 seconds.\n",
      "Running blast search for P20366.fasta...\n",
      "Blast search for P20366.fasta completed in 3097.21 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 06:30:35 GMT\n",
      "Running blast search for P35318.fasta...\n",
      "Blast search for P35318.fasta completed in 247.97 seconds.\n",
      "Running blast search for P12872.fasta...\n",
      "Blast search for P12872.fasta completed in 330.19 seconds.\n",
      "Running blast search for P10645.fasta...\n",
      "Blast search for P10645.fasta completed in 1289.84 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 07:01:43 GMT\n",
      "Running blast search for O95760.fasta...\n",
      "Blast search for O95760.fasta completed in 988.63 seconds.\n",
      "Running blast search for P01344.fasta...\n",
      "Blast search for P01344.fasta completed in 755.54 seconds.\n",
      "Running blast search for O14879.fasta...\n",
      "Blast search for O14879.fasta completed in 199.85 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 07:34:07 GMT\n",
      "Running blast search for O43745.fasta...\n",
      "Blast search for O43745.fasta completed in 758.55 seconds.\n",
      "Running blast search for Q92968.fasta...\n",
      "Blast search for Q92968.fasta completed in 138.22 seconds.\n",
      "Running blast search for O95390.fasta...\n",
      "Blast search for O95390.fasta completed in 330.19 seconds.\n",
      "Running blast search for P02533.fasta...\n",
      "Blast search for P02533.fasta completed in 187.50 seconds.\n",
      "Running blast search for Q01629.fasta...\n",
      "Blast search for Q01629.fasta completed in 437.44 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 08:04:59 GMT\n",
      "Running blast search for P0C0P6.fasta...\n",
      "Blast search for P0C0P6.fasta completed in 248.07 seconds.\n",
      "Running blast search for Q00534.fasta...\n",
      "Blast search for Q00534.fasta completed in 146.93 seconds.\n",
      "Running blast search for P20592.fasta...\n",
      "Blast search for P20592.fasta completed in 250.90 seconds.\n",
      "Running blast search for P06493.fasta...\n",
      "Blast search for P06493.fasta completed in 447.08 seconds.\n",
      "Running blast search for P06400.fasta...\n",
      "Blast search for P06400.fasta completed in 185.05 seconds.\n",
      "Running blast search for Q03252.fasta...\n",
      "Blast search for Q03252.fasta completed in 437.21 seconds.\n",
      "Running blast search for P24941.fasta...\n",
      "Blast search for P24941.fasta completed in 340.10 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 08:39:15 GMT\n",
      "Running blast search for Q13519.fasta...\n",
      "Blast search for Q13519.fasta completed in 437.36 seconds.\n",
      "Running blast search for P01574.fasta...\n",
      "Blast search for P01574.fasta completed in 248.26 seconds.\n",
      "Running blast search for P19883.fasta...\n",
      "Blast search for P19883.fasta completed in 138.54 seconds.\n",
      "Running blast search for P01282.fasta...\n",
      "Blast search for P01282.fasta completed in 1289.84 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 09:14:29 GMT\n",
      "Running blast search for Q9Y5Q6.fasta...\n",
      "Blast search for Q9Y5Q6.fasta completed in 1290.20 seconds.\n",
      "Running blast search for Q8IUX4.fasta...\n",
      "Blast search for Q8IUX4.fasta completed in 989.43 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 09:52:28 GMT\n",
      "Running blast search for P55265.fasta...\n",
      "Blast search for P55265.fasta completed in 71.43 seconds.\n",
      "Running blast search for Q99988.fasta...\n",
      "Blast search for Q99988.fasta completed in 248.16 seconds.\n",
      "Running blast search for P48645.fasta...\n",
      "Blast search for P48645.fasta completed in 437.33 seconds.\n",
      "Running blast search for P06396.fasta...\n",
      "Blast search for P06396.fasta completed in 195.49 seconds.\n",
      "Running blast search for P01966.fasta...\n",
      "Blast search for P01966.fasta completed in 138.48 seconds.\n",
      "Running blast search for Q96BS2.fasta...\n",
      "Blast search for Q96BS2.fasta completed in 250.84 seconds.\n",
      "Running blast search for P47898.fasta...\n",
      "Blast search for P47898.fasta completed in 330.38 seconds.\n",
      "Running blast search for P42226.fasta...\n",
      "Blast search for P42226.fasta completed in 333.03 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 10:25:53 GMT\n",
      "Running blast search for P09914.fasta...\n",
      "Blast search for P09914.fasta completed in 262.19 seconds.\n",
      "Running blast search for O75192.fasta...\n",
      "Blast search for O75192.fasta completed in 248.06 seconds.\n",
      "Running blast search for O95399.fasta...\n",
      "Blast search for O95399.fasta completed in 330.74 seconds.\n",
      "Running blast search for P11802.fasta...\n",
      "Blast search for P11802.fasta completed in 447.72 seconds.\n",
      "Running blast search for P08949.fasta...\n",
      "Blast search for P08949.fasta completed in 184.98 seconds.\n",
      "Running blast search for Q01453.fasta...\n",
      "Blast search for Q01453.fasta completed in 437.53 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 10:57:45 GMT\n",
      "Running blast search for Q10589.fasta...\n",
      "Blast search for Q10589.fasta completed in 989.06 seconds.\n",
      "Running blast search for Q969E3.fasta...\n",
      "Blast search for Q969E3.fasta completed in 135.82 seconds.\n",
      "Running blast search for Q14790.fasta...\n",
      "Blast search for Q14790.fasta completed in 136.01 seconds.\n",
      "Running blast search for A0A097EV82.fasta...\n",
      "Blast search for A0A097EV82.fasta completed in 988.81 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 11:35:14 GMT\n",
      "Running blast search for P55089.fasta...\n",
      "Blast search for P55089.fasta completed in 988.77 seconds.\n",
      "Running blast search for O75381.fasta...\n",
      "Blast search for O75381.fasta completed in 185.88 seconds.\n",
      "Running blast search for Q9P0W0.fasta...\n",
      "Blast search for Q9P0W0.fasta completed in 330.39 seconds.\n",
      "Running blast search for Q9H293.fasta...\n",
      "Blast search for Q9H293.fasta completed in 755.89 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 12:12:55 GMT\n",
      "Running blast search for Q00535.fasta...\n",
      "Blast search for Q00535.fasta completed in 1300.07 seconds.\n",
      "Running blast search for P43490.fasta...\n",
      "Blast search for P43490.fasta completed in 1297.22 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 12:56:13 GMT\n",
      "Running blast search for Q9BTW9.fasta...\n",
      "Blast search for Q9BTW9.fasta completed in 100.35 seconds.\n",
      "Running blast search for P42224.fasta...\n",
      "Blast search for P42224.fasta completed in 250.92 seconds.\n",
      "Running blast search for Q00526.fasta...\n",
      "Blast search for Q00526.fasta completed in 446.71 seconds.\n",
      "Running blast search for Q9HC23.fasta...\n",
      "Blast search for Q9HC23.fasta completed in 575.58 seconds.\n",
      "Running blast search for P01241.fasta...\n",
      "Blast search for P01241.fasta completed in 331.92 seconds.\n",
      "Running blast search for Q16611.fasta...\n",
      "Blast search for Q16611.fasta completed in 331.58 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 13:30:10 GMT\n",
      "Running blast search for P56199.fasta...\n",
      "Blast search for P56199.fasta completed in 262.71 seconds.\n",
      "Running blast search for P01258.fasta...\n",
      "Blast search for P01258.fasta completed in 330.32 seconds.\n",
      "Running blast search for Q8NG41.fasta...\n",
      "Blast search for Q8NG41.fasta completed in 135.83 seconds.\n",
      "Running blast search for P10082.fasta...\n",
      "Blast search for P10082.fasta completed in 248.38 seconds.\n",
      "Running blast search for Q7Z2W4.fasta...\n",
      "Blast search for Q7Z2W4.fasta completed in 330.27 seconds.\n",
      "Running blast search for P19525.fasta...\n",
      "Blast search for P19525.fasta completed in 257.89 seconds.\n",
      "Running blast search for O43808.fasta...\n",
      "Blast search for O43808.fasta completed in 138.82 seconds.\n",
      "Running blast search for Q96RP3.fasta...\n",
      "Blast search for Q96RP3.fasta completed in 330.25 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 14:04:04 GMT\n",
      "Running blast search for P01566.fasta...\n",
      "Blast search for P01566.fasta completed in 185.63 seconds.\n",
      "Running blast search for P01568.fasta...\n",
      "Blast search for P01568.fasta completed in 330.57 seconds.\n",
      "Running blast search for P80303.fasta...\n",
      "Blast search for P80303.fasta completed in 137.57 seconds.\n",
      "Running blast search for O15130.fasta...\n",
      "Blast search for O15130.fasta completed in 575.46 seconds.\n",
      "Running blast search for P01019.fasta...\n",
      "Blast search for P01019.fasta completed in 999.60 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 14:41:13 GMT\n",
      "Running blast search for Q99653.fasta...\n",
      "Blast search for Q99653.fasta completed in 187.47 seconds.\n",
      "Running blast search for P01570.fasta...\n",
      "Blast search for P01570.fasta completed in 185.20 seconds.\n",
      "Running blast search for P30990.fasta...\n",
      "Blast search for P30990.fasta completed in 184.85 seconds.\n",
      "Running blast search for P23582.fasta...\n",
      "Blast search for P23582.fasta completed in 250.43 seconds.\n",
      "Running blast search for P42574.fasta...\n",
      "Blast search for P42574.fasta completed in 330.26 seconds.\n",
      "Running blast search for P16860.fasta...\n",
      "Blast search for P16860.fasta completed in 330.44 seconds.\n",
      "Running blast search for Q14765.fasta...\n",
      "Blast search for Q14765.fasta completed in 138.31 seconds.\n",
      "Running blast search for Q9Y3Z3.fasta...\n",
      "Blast search for Q9Y3Z3.fasta completed in 259.12 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 15:12:19 GMT\n",
      "Running blast search for P01579.fasta...\n",
      "Blast search for P01579.fasta completed in 98.14 seconds.\n",
      "Running blast search for Q02388.fasta...\n",
      "Blast search for Q02388.fasta completed in 267.22 seconds.\n",
      "Running blast search for Q8N729.fasta...\n",
      "Blast search for Q8N729.fasta completed in 184.93 seconds.\n",
      "Running blast search for O95786.fasta...\n",
      "Blast search for O95786.fasta completed in 195.71 seconds.\n",
      "Running blast search for P13164.fasta...\n",
      "Blast search for P13164.fasta completed in 991.10 seconds.\n",
      "Running blast search for P18509.fasta...\n",
      "Blast search for P18509.fasta completed in 136.33 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 15:43:33 GMT\n",
      "Running blast search for Q12965.fasta...\n",
      "Blast search for Q12965.fasta completed in 251.35 seconds.\n",
      "Running blast search for P01160.fasta...\n",
      "Blast search for P01160.fasta completed in 97.67 seconds.\n",
      "Running blast search for P81605.fasta...\n",
      "Blast search for P81605.fasta completed in 1289.70 seconds.\n",
      "Running blast search for P09912.fasta...\n",
      "Blast search for P09912.fasta completed in 437.31 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 16:18:09 GMT\n",
      "Running blast search for Q13325.fasta...\n",
      "Blast search for Q13325.fasta completed in 144.46 seconds.\n",
      "Running blast search for P20591.fasta...\n",
      "Blast search for P20591.fasta completed in 332.70 seconds.\n",
      "Running blast search for P52630.fasta...\n",
      "Blast search for P52630.fasta completed in 100.10 seconds.\n",
      "Running blast search for P01210.fasta...\n",
      "Blast search for P01210.fasta completed in 68.55 seconds.\n",
      "Running blast search for P40763.fasta...\n",
      "Blast search for P40763.fasta completed in 187.47 seconds.\n",
      "Running blast search for P01213.fasta...\n",
      "Blast search for P01213.fasta completed in 437.52 seconds.\n",
      "Running blast search for Q01628.fasta...\n",
      "Blast search for Q01628.fasta completed in 988.77 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 16:55:48 GMT\n",
      "Running blast search for P23142.fasta...\n",
      "Blast search for P23142.fasta completed in 100.08 seconds.\n",
      "Running blast search for Q9UHF0.fasta...\n",
      "Blast search for Q9UHF0.fasta completed in 755.90 seconds.\n",
      "Running blast search for P51587.fasta...\n",
      "Blast search for P51587.fasta completed in 437.76 seconds.\n",
      "Running blast search for P15502.fasta...\n",
      "Blast search for P15502.fasta completed in 580.47 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 17:27:02 GMT\n",
      "Running blast search for P19838.fasta...\n",
      "Blast search for P19838.fasta completed in 268.38 seconds.\n",
      "Running blast search for P04637.fasta...\n",
      "Blast search for P04637.fasta completed in 138.38 seconds.\n",
      "Running blast search for Q5H8A3.fasta...\n",
      "Blast search for Q5H8A3.fasta completed in 248.58 seconds.\n",
      "Running blast search for P55211.fasta...\n",
      "Blast search for P55211.fasta completed in 248.06 seconds.\n",
      "Running blast search for P25445.fasta...\n",
      "Blast search for P25445.fasta completed in 1590.82 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-04 18:08:37 GMT\n",
      "Running blast search for Q9Y5Y5.fasta...\n",
      "Blast search for Q9Y5Y5.fasta completed in 188.26 seconds.\n",
      "Running blast search for P01303.fasta...\n",
      "Blast search for P01303.fasta completed in 755.62 seconds.\n",
      "Running blast search for P55210.fasta...\n",
      "Blast search for P55210.fasta completed in 184.84 seconds.\n",
      "Running blast search for P01225.fasta...\n",
      "Blast search for P01225.fasta completed in 135.81 seconds.\n",
      "Blastp searches completed for all fasta files.\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "from datetime import datetime, timezone\n",
    "\n",
    "# Path to the directory containing fasta files\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta\"\n",
    "\n",
    "# Directory to store blast results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts\"\n",
    "\n",
    "# Create the blast output directory if it doesn't exist\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# Record the start time of the entire process in GMT\n",
    "start_time = datetime.now(timezone.utc)\n",
    "print(f\"Process started at: {start_time.strftime('%Y-%m-%d %H:%M:%S')} GMT\")\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):  \n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        output_file = os.path.join(blast_output_directory, f\"{os.path.splitext(filename)[0]}_blastp_results.txt\")\n",
    "        \n",
    "        # Skip if output file already exists\n",
    "        if os.path.exists(output_file):\n",
    "            print(f\"Skipping {filename} as output file already exists.\")\n",
    "            continue\n",
    "        \n",
    "        # Construct the blastp command\n",
    "        blastp_command = [\n",
    "            \"blastp\",\n",
    "            \"-db\",\n",
    "            \"nr\",\n",
    "            \"-query\",\n",
    "            fasta_file,\n",
    "            \"-out\",\n",
    "            output_file,\n",
    "            \"-evalue\",\n",
    "            \"0.06\",\n",
    "            \"-remote\",\n",
    "            \"-entrez_query\",\n",
    "            \"archaea[orgn]\"\n",
    "        ]\n",
    "        \n",
    "        # Record start time for this file\n",
    "        file_start_time = time.time()\n",
    "        \n",
    "        # Execute the blastp command\n",
    "        print(f\"Running blast search for {filename}...\")\n",
    "        subprocess.run(blastp_command)\n",
    "        \n",
    "        # Calculate elapsed time for this file\n",
    "        file_elapsed_time = time.time() - file_start_time\n",
    "        \n",
    "        # Print filename and elapsed time\n",
    "        print(f\"Blast search for {filename} completed in {file_elapsed_time:.2f} seconds.\")\n",
    "        \n",
    "        # Check if 30 minutes have elapsed since the start of the process\n",
    "        if (datetime.now(timezone.utc) - start_time).seconds >= 1800:\n",
    "            print(f\"Process has been running for 30 minutes. Current time: {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M:%S')} GMT\")\n",
    "            start_time = datetime.now(timezone.utc)  \n",
    "\n",
    "print(\"Blastp searches completed for all fasta files.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Create Archaeal Master File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All blast_results.txt files have been concatenated into the master file: /home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts/archaeal_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Specify Directory containing blast results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts\"\n",
    "\n",
    "# Name the file\n",
    "master_filename = \"archaeal_master_blaster.txt\"\n",
    "\n",
    "# create Path to the master file\n",
    "master_filepath = os.path.join(blast_output_directory, master_filename)\n",
    "\n",
    "# Open the master file in append mode\n",
    "with open(master_filepath, \"a\") as master_file:\n",
    "    # Iterate through each file in the blast output directory\n",
    "    for filename in os.listdir(blast_output_directory):\n",
    "        # Check if the file is a blast result file\n",
    "        if filename.endswith(\"_blast_results.txt\"):\n",
    "            # Get the path of the blast result file\n",
    "            blast_result_filepath = os.path.join(blast_output_directory, filename)\n",
    "            # Open the blast result file\n",
    "            with open(blast_result_filepath, \"r\") as blast_result_file:\n",
    "                # Write contents to master file\n",
    "                master_file.write(blast_result_file.read())\n",
    "                # Separate the content of each blast result file\n",
    "                master_file.write(\"\\n\")\n",
    "\n",
    "print(\"All blast_results.txt files have been concatenated into the master file:\", master_filepath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ranked total hits for each accession ID:\n",
      "An error occurred while writing to CSV: [Errno 2] No such file or directory: ' /home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts/archaeal_master_blaster.csv'\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import csv\n",
    "\n",
    "# Path to the master file\n",
    "master_filepath = \"/home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts/archaeal_master_blaster.txt\"\n",
    "\n",
    "# Initialize a defaultdict to store query counts and hits\n",
    "query_hits = defaultdict(list)\n",
    "\n",
    "# Read the file and extract query IDs and hits\n",
    "try:\n",
    "    with open(master_filepath, \"r\") as master_file:\n",
    "        current_query_id = None\n",
    "        for line in master_file:\n",
    "            if \"|\" in line: \n",
    "                current_query_id = line.split(\"|\")[1].strip()  \n",
    "            elif \">\" in line: \n",
    "                if current_query_id:\n",
    "                    query_hits[current_query_id].append(line.split(\">\")[1].strip())  \n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: File '{master_filepath}' not found.\")\n",
    "    exit(1)\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {e}\")\n",
    "    exit(1)\n",
    "\n",
    "# Calculate total hits for each accession ID\n",
    "total_hits = {query_id: len(hits) for query_id, hits in query_hits.items()}\n",
    "\n",
    "# Rank and print total hits\n",
    "print(\"Ranked total hits for each accession ID:\")\n",
    "for rank, (query_id, hits) in enumerate(sorted(total_hits.items(), key=lambda x: x[1], reverse=True), start=1):\n",
    "    print(f\"{rank}. Accession ID: {query_id}, Total Hits: {hits}\")\n",
    "\n",
    "# Write results to CSV file\n",
    "output_csv_path = \" /home/osheakes/Research_Project_MMM/Fasta/Archaeal_Blasts/archaeal_master_blaster.csv\"\n",
    "try:\n",
    "    with open(output_csv_path, mode='w', newline='') as csv_file:\n",
    "        fieldnames = ['Query ID', 'Hits']\n",
    "        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)\n",
    "        \n",
    "        writer.writeheader()\n",
    "        for query_id, hits in query_hits.items():\n",
    "            writer.writerow({'Query ID': query_id, 'Hits': ', '.join(hits)})\n",
    "    print(f\"Results saved to '{output_csv_path}'\")\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred while writing to CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.0 Fungal Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /home/osheakes/Research_Project_MMM/Fasta/Fungal_Blasts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process started at: 2024-05-05 02:31:31 GMT\n",
      "Running blast search for O96011.fasta...\n",
      "Blast search for O96011.fasta completed in 149.20 seconds.\n",
      "Running blast search for Q07812.fasta...\n",
      "Blast search for Q07812.fasta completed in 184.94 seconds.\n",
      "Running blast search for Q8NFJ6.fasta...\n",
      "Blast search for Q8NFJ6.fasta completed in 250.96 seconds.\n",
      "Running blast search for P07492.fasta...\n",
      "Blast search for P07492.fasta completed in 186.15 seconds.\n",
      "Running blast search for Q15303.fasta...\n",
      "Blast search for Q15303.fasta completed in 195.64 seconds.\n",
      "Running blast search for Q15726.fasta...\n",
      "Blast search for Q15726.fasta completed in 97.84 seconds.\n",
      "Running blast search for O00910.fasta...\n",
      "Blast search for O00910.fasta completed in 184.86 seconds.\n",
      "Running blast search for P38398.fasta...\n",
      "Blast search for P38398.fasta completed in 766.30 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 03:05:07 GMT\n",
      "Running blast search for P06881.fasta...\n",
      "Blast search for P06881.fasta completed in 185.03 seconds.\n",
      "Running blast search for P46663.fasta...\n",
      "Blast search for P46663.fasta completed in 138.30 seconds.\n",
      "Running blast search for P32247.fasta...\n",
      "Blast search for P32247.fasta completed in 187.92 seconds.\n",
      "Running blast search for P60484.fasta...\n",
      "Blast search for P60484.fasta completed in 258.58 seconds.\n",
      "Running blast search for P20366.fasta...\n",
      "Blast search for P20366.fasta completed in 185.03 seconds.\n",
      "Running blast search for P35318.fasta...\n",
      "Blast search for P35318.fasta completed in 248.14 seconds.\n",
      "Running blast search for P12872.fasta...\n",
      "Blast search for P12872.fasta completed in 135.74 seconds.\n",
      "Running blast search for P10645.fasta...\n",
      "Blast search for P10645.fasta completed in 248.15 seconds.\n",
      "Running blast search for O95760.fasta...\n",
      "Blast search for O95760.fasta completed in 248.06 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 03:35:42 GMT\n",
      "Running blast search for P01344.fasta...\n",
      "Blast search for P01344.fasta completed in 135.72 seconds.\n",
      "Running blast search for O14879.fasta...\n",
      "Blast search for O14879.fasta completed in 190.97 seconds.\n",
      "Running blast search for O43745.fasta...\n",
      "Blast search for O43745.fasta completed in 259.73 seconds.\n",
      "Running blast search for Q92968.fasta...\n",
      "Blast search for Q92968.fasta completed in 196.55 seconds.\n",
      "Running blast search for O95390.fasta...\n",
      "Blast search for O95390.fasta completed in 137.31 seconds.\n",
      "Running blast search for P02533.fasta...\n",
      "Blast search for P02533.fasta completed in 187.53 seconds.\n",
      "Running blast search for Q01629.fasta...\n",
      "Blast search for Q01629.fasta completed in 184.84 seconds.\n",
      "Running blast search for P0C0P6.fasta...\n",
      "Blast search for P0C0P6.fasta completed in 185.56 seconds.\n",
      "Running blast search for Q00534.fasta...\n",
      "Blast search for Q00534.fasta completed in 258.89 seconds.\n",
      "Running blast search for P20592.fasta...\n",
      "Blast search for P20592.fasta completed in 146.74 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 04:07:06 GMT\n",
      "Running blast search for P06493.fasta...\n",
      "Blast search for P06493.fasta completed in 194.93 seconds.\n",
      "Running blast search for P06400.fasta...\n",
      "Blast search for P06400.fasta completed in 100.74 seconds.\n",
      "Running blast search for Q03252.fasta...\n",
      "Blast search for Q03252.fasta completed in 187.82 seconds.\n",
      "Running blast search for P24941.fasta...\n",
      "Blast search for P24941.fasta completed in 194.93 seconds.\n",
      "Running blast search for Q13519.fasta...\n",
      "Blast search for Q13519.fasta completed in 97.76 seconds.\n",
      "Running blast search for P01574.fasta...\n",
      "Blast search for P01574.fasta completed in 71.18 seconds.\n",
      "Running blast search for P19883.fasta...\n",
      "Blast search for P19883.fasta completed in 72.57 seconds.\n",
      "Running blast search for P01282.fasta...\n",
      "Blast search for P01282.fasta completed in 97.58 seconds.\n",
      "Running blast search for Q9Y5Q6.fasta...\n",
      "Blast search for Q9Y5Q6.fasta completed in 756.10 seconds.\n",
      "Running blast search for Q8IUX4.fasta...\n",
      "Blast search for Q8IUX4.fasta completed in 184.91 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 04:39:44 GMT\n",
      "Running blast search for P55265.fasta...\n",
      "Blast search for P55265.fasta completed in 258.87 seconds.\n",
      "Running blast search for Q99988.fasta...\n",
      "Blast search for Q99988.fasta completed in 575.22 seconds.\n",
      "Running blast search for P48645.fasta...\n",
      "Blast search for P48645.fasta completed in 135.86 seconds.\n",
      "Running blast search for P06396.fasta...\n",
      "Blast search for P06396.fasta completed in 152.11 seconds.\n",
      "Running blast search for P01966.fasta...\n",
      "Blast search for P01966.fasta completed in 100.57 seconds.\n",
      "Running blast search for Q96BS2.fasta...\n",
      "Blast search for Q96BS2.fasta completed in 145.87 seconds.\n",
      "Running blast search for P47898.fasta...\n",
      "Blast search for P47898.fasta completed in 250.92 seconds.\n",
      "Running blast search for P42226.fasta...\n",
      "Blast search for P42226.fasta completed in 100.01 seconds.\n",
      "Running blast search for P09914.fasta...\n",
      "Blast search for P09914.fasta completed in 105.30 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 05:10:09 GMT\n",
      "Running blast search for O75192.fasta...\n",
      "Blast search for O75192.fasta completed in 107.69 seconds.\n",
      "Running blast search for O95399.fasta...\n",
      "Blast search for O95399.fasta completed in 185.17 seconds.\n",
      "Running blast search for P11802.fasta...\n",
      "Blast search for P11802.fasta completed in 78.51 seconds.\n",
      "Running blast search for P08949.fasta...\n",
      "Blast search for P08949.fasta completed in 68.91 seconds.\n",
      "Running blast search for Q01453.fasta...\n",
      "Blast search for Q01453.fasta completed in 135.71 seconds.\n",
      "Running blast search for Q10589.fasta...\n",
      "Blast search for Q10589.fasta completed in 100.68 seconds.\n",
      "Running blast search for Q969E3.fasta...\n",
      "Blast search for Q969E3.fasta completed in 68.61 seconds.\n",
      "Running blast search for Q14790.fasta...\n",
      "Blast search for Q14790.fasta completed in 97.74 seconds.\n",
      "Running blast search for A0A097EV82.fasta...\n",
      "Blast search for A0A097EV82.fasta completed in 97.72 seconds.\n",
      "Running blast search for P55089.fasta...\n",
      "Blast search for P55089.fasta completed in 97.64 seconds.\n",
      "Running blast search for O75381.fasta...\n",
      "Blast search for O75381.fasta completed in 146.44 seconds.\n",
      "Running blast search for Q9P0W0.fasta...\n",
      "Blast search for Q9P0W0.fasta completed in 101.32 seconds.\n",
      "Running blast search for Q9H293.fasta...\n",
      "Blast search for Q9H293.fasta completed in 101.13 seconds.\n",
      "Running blast search for Q00535.fasta...\n",
      "Blast search for Q00535.fasta completed in 146.47 seconds.\n",
      "Running blast search for P43490.fasta...\n",
      "Blast search for P43490.fasta completed in 77.47 seconds.\n",
      "Running blast search for Q9BTW9.fasta...\n",
      "Blast search for Q9BTW9.fasta completed in 108.93 seconds.\n",
      "Running blast search for P42224.fasta...\n",
      "Blast search for P42224.fasta completed in 578.13 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 05:48:27 GMT\n",
      "Running blast search for Q00526.fasta...\n",
      "Blast search for Q00526.fasta completed in 108.69 seconds.\n",
      "Running blast search for Q9HC23.fasta...\n",
      "Blast search for Q9HC23.fasta completed in 135.86 seconds.\n",
      "Running blast search for P01241.fasta...\n",
      "Blast search for P01241.fasta completed in 97.61 seconds.\n",
      "Running blast search for Q16611.fasta...\n",
      "Blast search for Q16611.fasta completed in 97.95 seconds.\n",
      "Running blast search for P56199.fasta...\n",
      "Blast search for P56199.fasta completed in 101.30 seconds.\n",
      "Running blast search for P01258.fasta...\n",
      "Blast search for P01258.fasta completed in 135.75 seconds.\n",
      "Running blast search for Q8NG41.fasta...\n",
      "Blast search for Q8NG41.fasta completed in 98.32 seconds.\n",
      "Running blast search for P10082.fasta...\n",
      "Blast search for P10082.fasta completed in 330.72 seconds.\n",
      "Running blast search for Q7Z2W4.fasta...\n",
      "Blast search for Q7Z2W4.fasta completed in 440.44 seconds.\n",
      "Running blast search for P19525.fasta...\n",
      "Blast search for P19525.fasta completed in 450.60 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 06:21:45 GMT\n",
      "Running blast search for O43808.fasta...\n",
      "Blast search for O43808.fasta completed in 147.02 seconds.\n",
      "Running blast search for Q96RP3.fasta...\n",
      "Blast search for Q96RP3.fasta completed in 184.88 seconds.\n",
      "Running blast search for P01566.fasta...\n",
      "Blast search for P01566.fasta completed in 99.97 seconds.\n",
      "Running blast search for P01568.fasta...\n",
      "Blast search for P01568.fasta completed in 137.98 seconds.\n",
      "Running blast search for P80303.fasta...\n",
      "Blast search for P80303.fasta completed in 104.06 seconds.\n",
      "Running blast search for O15130.fasta...\n",
      "Blast search for O15130.fasta completed in 184.81 seconds.\n",
      "Running blast search for P01019.fasta...\n",
      "Blast search for P01019.fasta completed in 100.06 seconds.\n",
      "Running blast search for Q99653.fasta...\n",
      "Blast search for Q99653.fasta completed in 194.59 seconds.\n",
      "Running blast search for P01570.fasta...\n",
      "Blast search for P01570.fasta completed in 100.06 seconds.\n",
      "Running blast search for P30990.fasta...\n",
      "Blast search for P30990.fasta completed in 247.93 seconds.\n",
      "Running blast search for P23582.fasta...\n",
      "Blast search for P23582.fasta completed in 330.82 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 06:52:17 GMT\n",
      "Running blast search for P42574.fasta...\n",
      "Blast search for P42574.fasta completed in 330.14 seconds.\n",
      "Running blast search for P16860.fasta...\n",
      "Blast search for P16860.fasta completed in 248.05 seconds.\n",
      "Running blast search for Q14765.fasta...\n",
      "Blast search for Q14765.fasta completed in 334.35 seconds.\n",
      "Running blast search for Q9Y3Z3.fasta...\n",
      "Blast search for Q9Y3Z3.fasta completed in 341.41 seconds.\n",
      "Running blast search for P01579.fasta...\n",
      "Blast search for P01579.fasta completed in 248.67 seconds.\n",
      "Running blast search for Q02388.fasta...\n",
      "Blast search for Q02388.fasta completed in 293.80 seconds.\n",
      "Running blast search for Q8N729.fasta...\n",
      "Blast search for Q8N729.fasta completed in 185.21 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 07:25:18 GMT\n",
      "Running blast search for O95786.fasta...\n",
      "Blast search for O95786.fasta completed in 341.26 seconds.\n",
      "Running blast search for P13164.fasta...\n",
      "Blast search for P13164.fasta completed in 330.15 seconds.\n",
      "Running blast search for P18509.fasta...\n",
      "Blast search for P18509.fasta completed in 135.79 seconds.\n",
      "Running blast search for Q12965.fasta...\n",
      "Blast search for Q12965.fasta completed in 586.81 seconds.\n",
      "Running blast search for P01160.fasta...\n",
      "Blast search for P01160.fasta completed in 330.11 seconds.\n",
      "Running blast search for P81605.fasta...\n",
      "Blast search for P81605.fasta completed in 330.33 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 07:59:33 GMT\n",
      "Running blast search for P09912.fasta...\n",
      "Blast search for P09912.fasta completed in 251.30 seconds.\n",
      "Running blast search for Q13325.fasta...\n",
      "Blast search for Q13325.fasta completed in 578.42 seconds.\n",
      "Running blast search for P20591.fasta...\n",
      "Blast search for P20591.fasta completed in 340.25 seconds.\n",
      "Running blast search for P52630.fasta...\n",
      "Blast search for P52630.fasta completed in 758.08 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 08:31:41 GMT\n",
      "Running blast search for P01210.fasta...\n",
      "Blast search for P01210.fasta completed in 755.67 seconds.\n",
      "Running blast search for P40763.fasta...\n",
      "Blast search for P40763.fasta completed in 577.99 seconds.\n",
      "Running blast search for P01213.fasta...\n",
      "Blast search for P01213.fasta completed in 184.87 seconds.\n",
      "Running blast search for Q01628.fasta...\n",
      "Blast search for Q01628.fasta completed in 184.93 seconds.\n",
      "Running blast search for P23142.fasta...\n",
      "Blast search for P23142.fasta completed in 187.99 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 09:03:12 GMT\n",
      "Running blast search for Q9UHF0.fasta...\n",
      "Blast search for Q9UHF0.fasta completed in 135.79 seconds.\n",
      "Running blast search for P51587.fasta...\n",
      "Blast search for P51587.fasta completed in 195.62 seconds.\n",
      "Running blast search for P15502.fasta...\n",
      "Blast search for P15502.fasta completed in 250.40 seconds.\n",
      "Running blast search for P19838.fasta...\n",
      "Blast search for P19838.fasta completed in 606.81 seconds.\n",
      "Running blast search for P04637.fasta...\n",
      "Blast search for P04637.fasta completed in 1289.60 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 09:44:31 GMT\n",
      "Running blast search for Q5H8A3.fasta...\n",
      "Blast search for Q5H8A3.fasta completed in 1591.23 seconds.\n",
      "Running blast search for P55211.fasta...\n",
      "Blast search for P55211.fasta completed in 437.40 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 10:18:19 GMT\n",
      "Running blast search for P25445.fasta...\n",
      "Blast search for P25445.fasta completed in 575.40 seconds.\n",
      "Running blast search for Q9Y5Y5.fasta...\n",
      "Blast search for Q9Y5Y5.fasta completed in 2203.77 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 11:04:38 GMT\n",
      "Running blast search for P01303.fasta...\n",
      "Blast search for P01303.fasta completed in 1289.80 seconds.\n",
      "Running blast search for P55210.fasta...\n",
      "Blast search for P55210.fasta completed in 1591.52 seconds.\n",
      "Process has been running for 30 minutes. Current time: 2024-05-05 11:52:40 GMT\n",
      "Running blast search for P01225.fasta...\n",
      "Blast search for P01225.fasta completed in 1591.69 seconds.\n",
      "Blastp searches completed for all fasta files.\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "from datetime import datetime, timezone\n",
    "\n",
    "# Path to the directory containing fasta files\n",
    "fasta_directory = \"/home/osheakes/Research_Project_MMM/Fasta\"\n",
    "\n",
    "# Directory to store blast results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Fungal_Blasts\"\n",
    "\n",
    "# Create the blast output directory if it doesn't exist\n",
    "os.makedirs(blast_output_directory, exist_ok=True)\n",
    "\n",
    "# Record the start time of the entire process in GMT\n",
    "start_time = datetime.now(timezone.utc)\n",
    "print(f\"Process started at: {start_time.strftime('%Y-%m-%d %H:%M:%S')} GMT\")\n",
    "\n",
    "# Iterate through each fasta file in the directory\n",
    "for filename in os.listdir(fasta_directory):\n",
    "    if filename.endswith(\".fasta\"):  \n",
    "        fasta_file = os.path.join(fasta_directory, filename)\n",
    "        output_file = os.path.join(blast_output_directory, f\"{os.path.splitext(filename)[0]}_blastp_results.txt\")\n",
    "        \n",
    "        # Skip if output file already exists\n",
    "        if os.path.exists(output_file):\n",
    "            print(f\"Skipping {filename} as output file already exists.\")\n",
    "            continue\n",
    "        \n",
    "        # Construct the blastp command\n",
    "        blastp_command = [\n",
    "            \"blastp\",\n",
    "            \"-db\",\n",
    "            \"nr\",\n",
    "            \"-query\",\n",
    "            fasta_file,\n",
    "            \"-out\",\n",
    "            output_file,\n",
    "            \"-evalue\",\n",
    "            \"0.06\",\n",
    "            \"-remote\",\n",
    "            \"-entrez_query\",\n",
    "            \"fungi[orgn]\"\n",
    "        ]\n",
    "        \n",
    "        # Record start time for this file\n",
    "        file_start_time = time.time()\n",
    "        \n",
    "        # Execute the blastp command\n",
    "        print(f\"Running blast search for {filename}...\")\n",
    "        subprocess.run(blastp_command)\n",
    "        \n",
    "        # Calculate elapsed time for this file\n",
    "        file_elapsed_time = time.time() - file_start_time\n",
    "        \n",
    "        # Print filename and elapsed time\n",
    "        print(f\"Blast search for {filename} completed in {file_elapsed_time:.2f} seconds.\")\n",
    "        \n",
    "        # Check if 30 minutes have elapsed since the start of the process\n",
    "        if (datetime.now(timezone.utc) - start_time).seconds >= 1800:\n",
    "            print(f\"Process has been running for 30 minutes. Current time: {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M:%S')} GMT\")\n",
    "            start_time = datetime.now(timezone.utc)  \n",
    "\n",
    "print(\"Blastp searches completed for all fasta files.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Create Fungal Master File "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All blast_results.txt files have been concatenated into the master file: /home/osheakes/Research_Project_MMM/Fasta/Fungal_Blasts/Fungal_master_blaster.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Specify Directory containing blast results\n",
    "blast_output_directory = \"/home/osheakes/Research_Project_MMM/Fasta/Fungal_Blasts\"\n",
    "\n",
    "# Name the file\n",
    "master_filename = \"Fungal_master_blaster.txt\"\n",
    "\n",
    "# create Path to the master file\n",
    "master_filepath = os.path.join(blast_output_directory, master_filename)\n",
    "\n",
    "# Open the master file in append mode\n",
    "with open(master_filepath, \"a\") as master_file:\n",
    "    # Iterate through each file in the blast output directory\n",
    "    for filename in os.listdir(blast_output_directory):\n",
    "        # Check if the file is a blast result file\n",
    "        if filename.endswith(\"_blast_results.txt\"):\n",
    "            # Get the path of the blast result file\n",
    "            blast_result_filepath = os.path.join(blast_output_directory, filename)\n",
    "            # Open the blast result file\n",
    "            with open(blast_result_filepath, \"r\") as blast_result_file:\n",
    "                # Write contents to master file\n",
    "                master_file.write(blast_result_file.read())\n",
    "                # Separate the content of each blast result file\n",
    "                master_file.write(\"\\n\")\n",
    "\n",
    "print(\"All blast_results.txt files have been concatenated into the master file:\", master_filepath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
